{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "25a3eb7b",
   "metadata": {},
   "source": [
    "### Load libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a9a52127",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Attaching SeuratObject\n",
      "\n",
      "Loading required package: viridisLite\n",
      "\n",
      "Loading required package: iterators\n",
      "\n",
      "Loading required package: parallel\n",
      "\n",
      "\n",
      "Attaching package: ‘proxy’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    as.dist, dist\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    as.matrix\n",
      "\n",
      "\n",
      "Loading required package: grid\n",
      "\n",
      "========================================\n",
      "ComplexHeatmap version 2.11.1\n",
      "Bioconductor page: http://bioconductor.org/packages/ComplexHeatmap/\n",
      "Github page: https://github.com/jokergoo/ComplexHeatmap\n",
      "Documentation: http://jokergoo.github.io/ComplexHeatmap-reference\n",
      "\n",
      "If you use it in published research, please cite:\n",
      "Gu, Z. Complex heatmaps reveal patterns and correlations in multidimensional \n",
      "  genomic data. Bioinformatics 2016.\n",
      "\n",
      "The new InteractiveComplexHeatmap package can directly export static \n",
      "complex heatmaps into an interactive Shiny app with zero effort. Have a try!\n",
      "\n",
      "This message can be suppressed by:\n",
      "  suppressPackageStartupMessages(library(ComplexHeatmap))\n",
      "========================================\n",
      "\n",
      "\n",
      "========================================\n",
      "circlize version 0.4.14\n",
      "CRAN page: https://cran.r-project.org/package=circlize\n",
      "Github page: https://github.com/jokergoo/circlize\n",
      "Documentation: https://jokergoo.github.io/circlize_book/book/\n",
      "\n",
      "If you use it in published research, please cite:\n",
      "Gu, Z. circlize implements and enhances circular visualization\n",
      "  in R. Bioinformatics 2014.\n",
      "\n",
      "This message can be suppressed by:\n",
      "  suppressPackageStartupMessages(library(circlize))\n",
      "========================================\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘igraph’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:circlize’:\n",
      "\n",
      "    degree\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    decompose, spectrum\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    union\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:igraph’:\n",
      "\n",
      "    as_data_frame, groups, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: splines\n",
      "\n",
      "\n",
      "Attaching package: ‘pheatmap’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:ComplexHeatmap’:\n",
      "\n",
      "    pheatmap\n",
      "\n",
      "\n",
      "Loading required package: princurve\n",
      "\n",
      "Loading required package: TrajectoryUtils\n",
      "\n",
      "Loading required package: SingleCellExperiment\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:igraph’:\n",
      "\n",
      "    normalize, path, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    first, rename\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    first, second\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    collapse, desc, slice\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:data.table’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘SummarizedExperiment’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SeuratObject’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Seurat’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(Seurat)\n",
    "library(stringr)\n",
    "library(viridis)\n",
    "library(ggplot2)\n",
    "library(cowplot)\n",
    "library(cluster)\n",
    "library(data.table)\n",
    "library(foreach)\n",
    "library(doParallel)\n",
    "library(proxy)\n",
    "library(ComplexHeatmap)\n",
    "library(circlize)\n",
    "library(igraph)\n",
    "library(qvalue)\n",
    "library(dplyr)\n",
    "library(viridis)\n",
    "library(VGAM)\n",
    "library(forcats)\n",
    "library(grDevices)\n",
    "library(graphics)\n",
    "library(RColorBrewer)\n",
    "library(pheatmap)\n",
    "library(Cairo)\n",
    "library(slingshot)\n",
    "\n",
    "createEmptyDf = function( nrow, ncol, colnames = c() ){\n",
    "  data.frame( matrix( vector(), nrow, ncol, dimnames = list( c(), colnames ) ) )\n",
    "}\n",
    "\n",
    "tableread_fast = function(i, header=TRUE, quote=\"\", sep=\",\"){\n",
    "  tmp = fread(i, header=header, sep=sep, quote=quote, nThread=32)\n",
    "  tmp = as.data.frame(tmp)\n",
    "  return(tmp)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f207b132",
   "metadata": {},
   "source": [
    "### Load functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "54558b75",
   "metadata": {},
   "outputs": [],
   "source": [
    "createEmptyDf = function( nrow, ncol, colnames = c() ){\n",
    "  data.frame( matrix( vector(), nrow, ncol, dimnames = list( c(), colnames ) ) )\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b9f8d6b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Stacked Violin Plot\n",
    "#https://divingintogeneticsandgenomics.rbind.io/post/stacked-violin-plot-for-visualizing-single-cell-data-in-seurat/\n",
    "\n",
    "## remove the x-axis text and tick\n",
    "## plot.margin to adjust the white space between each plot.\n",
    "## ... pass any arguments to VlnPlot in Seurat\n",
    "modify_vlnplot<- function(obj, \n",
    "                          feature, \n",
    "                          pt.size = 0, \n",
    "                          plot.margin = unit(c(-0.75, 0, -0.75, 0), \"cm\"),\n",
    "                          ...) {\n",
    "  p<- VlnPlot(obj, features = feature, pt.size = pt.size, ... )  + \n",
    "    xlab(\"\") + ylab(feature) + ggtitle(\"\") + \n",
    "    theme(legend.position = \"none\", \n",
    "          plot.title= element_blank(),\n",
    "          axis.title.x = element_blank(),\n",
    "          axis.text.x = element_blank(), \n",
    "          axis.ticks.x = element_blank(), \n",
    "          axis.title.y = element_text(family = \"Arial\", size = 8, angle = 0, colour = \"black\", face = \"plain\"), \n",
    "          axis.text.y = element_blank(), \n",
    "          plot.margin = plot.margin ) \n",
    "  return(p)\n",
    "}\n",
    "## extract the max value of the y axis\n",
    "extract_max<- function(p){\n",
    "  ymax<- max(ggplot_build(p)$layout$panel_scales_y[[1]]$range$range)\n",
    "  return(ceiling(ymax))\n",
    "}\n",
    "## main function\n",
    "StackedVlnPlot<- function(obj, features,\n",
    "                          pt.size = 0, \n",
    "                          plot.margin = unit(c(-0.05, 0, -0.05, 0), \"cm\"),\n",
    "                          ...) {\n",
    "  \n",
    "  plot_list<- purrr::map(features, function(x) modify_vlnplot(obj = obj,feature = x, ...))\n",
    "  \n",
    "  # Add back x-axis title to bottom plot. patchwork is going to support this?\n",
    "  plot_list[[length(plot_list)]]<- plot_list[[length(plot_list)]] +\n",
    "    theme(axis.text.x=element_text(family = \"Arial\", size=8, face = \"plain\"),\n",
    "          axis.ticks.x = element_line())\n",
    "  \n",
    "  \n",
    "  # change the y-axis tick to only max value \n",
    "  ymaxs<- purrr::map_dbl(plot_list, extract_max)\n",
    "  plot_list<- purrr::map2(plot_list, ymaxs, function(x,y) x + \n",
    "                            scale_y_continuous(breaks = c(y)) + \n",
    "                            expand_limits(y = y))\n",
    "  \n",
    "  p<- patchwork::wrap_plots(plotlist = plot_list, ncol = 1, heights = 1)\n",
    "  return(p)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "86afb98c",
   "metadata": {},
   "outputs": [],
   "source": [
    "modify_vlnplot_score <- function(obj, \n",
    "                          feature, \n",
    "                          pt.size = 0, \n",
    "                          plot.margin = unit(c(-0.75, 0, -0.75, 0), \"cm\"),\n",
    "                          ...) {\n",
    "  p<- VlnPlot(obj, features = feature, pt.size = pt.size, ... )  + \n",
    "    xlab(\"\") + ylab(feature) + ggtitle(\"\") + \n",
    "    theme(legend.position = \"none\", \n",
    "          plot.title= element_blank(),\n",
    "          axis.title.x = element_blank(),\n",
    "          axis.text.x =  element_text(family = \"Arial\", size = 8, angle = 0, colour = \"black\", face = \"plain\"), \n",
    "          axis.ticks.x = element_blank(), \n",
    "          axis.title.y = element_blank(), \n",
    "          axis.text.y = element_text(family = \"Arial\", size = 6, angle = 0, colour = \"black\", face = \"plain\"), \n",
    "          plot.margin = plot.margin ) \n",
    "  return(p)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "2e410617",
   "metadata": {},
   "outputs": [],
   "source": [
    "##Plot clones in dimentional reduction plot (DimPlot)\n",
    "#seu: Seurat object / tmp2: identifier of plotting cells\n",
    "\n",
    "PlotClones = function(seu, tmp2, name.output, meta.data, clone_plot, reduction){\n",
    "    #Exctract names for cells in plot\n",
    "    tmp_raw <- as.vector(row.names(meta.data))\n",
    "    #Extract colors for plot\n",
    "    p2 = Seurat::DimPlot(object = seu, reduction = reduction, pt.size = 0.1)\n",
    "    orig_ident_build = ggplot2::ggplot_build(p2)\n",
    "    orig_ident_build = orig_ident_build$data[[1]]\n",
    "    orig_ident_build =  orig_ident_build[order(orig_ident_build$group), ]\n",
    "    ident.cols = unique(orig_ident_build$colour) # Get a vector of unique colors\n",
    "    names(ident.cols)=c(0:(max(as.numeric(seu@active.ident)-1)))\n",
    "\n",
    "    #Extract x- and y- axis information for DimPlot\n",
    "    xmin <- min(orig_ident_build$x)-0.5\n",
    "    xmax <- max(orig_ident_build$x)+0.5\n",
    "    ymin <- min(orig_ident_build$y)-0.5\n",
    "    ymax <- max(orig_ident_build$y)+0.5\n",
    "    x_label_min <- floor(xmin/10)*10\n",
    "    x_label_max <- ceiling(xmax/10)*10\n",
    "    y_label_min <- floor(ymin/10)*10\n",
    "    y_label_max <- ceiling(ymax/10)*10\n",
    "\n",
    "    #Dimplot\n",
    "    p1 <- DimPlot(object = seu, label = FALSE,\n",
    "                    cells = tmp2,\n",
    "                    reduction = reduction,\n",
    "                    cols = ident.cols,\n",
    "                    pt.size = 0.1) + \n",
    "        NoLegend() +\n",
    "        scale_x_continuous(limits = c(xmin,xmax), breaks= seq(x_label_min,x_label_max,10)) +\n",
    "        scale_y_continuous(limits = c(ymin,ymax), breaks= seq(y_label_min,y_label_max,10)) +\n",
    "        theme(axis.title = element_blank(), \n",
    "              axis.text = element_blank(),\n",
    "             axis.line = element_line(size = 0.2))\n",
    "    \n",
    "      save_plot(file = name.output, plot_grid(p1, ncol=1, nrow=1), device=\"tiff\", \n",
    "                units=\"in\", dpi = 300, base_width = 3, base_height = 3, limitsize=FALSE)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "5220d0ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Pseudotime analysis\n",
    "#seu.sub: create pseudotime for subset of clusters\n",
    "#red.use: dimentional reduction for plotting pseudotime\n",
    "Pseudotime.Min <- function(seu, seu.sub, name, dims, start, dir.name){\n",
    "  #Create lineages and pseudotime\n",
    "  lineages = slingshot(seu.sub@reductions$pca@cell.embeddings[,dims], \n",
    "                       clusterLabels = seu.sub@active.ident, maxit=1000, start.clus=start)\n",
    "  pseudotime=slingPseudotime(lineages, na = TRUE)\n",
    "  \n",
    "  #calculate pseudotime for each lineage\n",
    "  for (i in c(1:length(lineages@metadata$lineages))){\n",
    "    pseudotime[,i]=pseudotime[,i]/max(pseudotime[,i][!is.na(pseudotime[,i])])\n",
    "  }\n",
    "  if (length(lineages@metadata$lineages) == 1) {\n",
    "    pseudotime1 = pseudotime[,1]\n",
    "  } else {\n",
    "    pseudotime1 = Matrix::rowMeans(pseudotime, na.rm=TRUE)\n",
    "  }\n",
    "  \n",
    "  #Add to MetaData\n",
    "  seu2 = AddMetaData(object = seu, metadata = pseudotime1, col.name = name)\n",
    "    \n",
    "  return(seu2)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2861f51d",
   "metadata": {},
   "source": [
    "### Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "5e6278b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "name.input <- \"Aoki_3rd_res0.6scTCRmerged.ABnt.AddSig.rda\"\n",
    "load(name.input)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02dbfe85",
   "metadata": {},
   "source": [
    "### Figure 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b24b6e13",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(dir.name):\n",
      "“ '3B' はすでに存在します ”\n"
     ]
    }
   ],
   "source": [
    "###Figure 3B, Protting the result of dimentional reduction\n",
    "\n",
    "dir.name <- \"3B\"\n",
    "file.name <- \"Fig3B.umap.tiff\"\n",
    "\n",
    "################################# Processing layer ##########################################\n",
    "dir.create(dir.name)\n",
    "\n",
    "#Output twice size of figure, compress after pasting in the powerpoint.\n",
    "p1 = DimPlot(object = seu, reduction = \"umap\", label = TRUE, label.size = 8, pt.size = 0.2) +\n",
    "  theme(axis.title.x = element_text(size=16, family = \"Arial\"), \n",
    "        axis.title.y = element_text(size=16, family = \"Arial\"), \n",
    "        axis.text.x = element_text(size=12, colour = 1, family = \"Arial\"), \n",
    "        axis.text.y = element_text(size = 12, colour = 1, family = \"Arial\")) +\n",
    "  theme(panel.border = element_rect(fill = NA, size = 1.5)) \n",
    "\n",
    "p1 = p1 + theme(legend.position = 'none')\n",
    "name.output <- str_c(dir.name, file.name, sep = \"/\")\n",
    "save_plot(file = name.output, plot(p1), device=\"tiff\", \n",
    "          units=\"in\", dpi = 300, base_width = 4.8, base_height = 4.8, limitsize=FALSE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "6b1a7237",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(dir.name):\n",
      "“ '3C' はすでに存在します ”\n",
      "Scale for 'y' is already present. Adding another scale for 'y', which will\n",
      "replace the existing scale.\n",
      "\n",
      "Scale for 'y' is already present. Adding another scale for 'y', which will\n",
      "replace the existing scale.\n",
      "\n",
      "Scale for 'y' is already present. Adding another scale for 'y', which will\n",
      "replace the existing scale.\n",
      "\n",
      "Scale for 'y' is already present. Adding another scale for 'y', which will\n",
      "replace the existing scale.\n",
      "\n",
      "Scale for 'y' is already present. Adding another scale for 'y', which will\n",
      "replace the existing scale.\n",
      "\n",
      "Scale for 'y' is already present. Adding another scale for 'y', which will\n",
      "replace the existing scale.\n",
      "\n",
      "Scale for 'y' is already present. Adding another scale for 'y', which will\n",
      "replace the existing scale.\n",
      "\n",
      "Scale for 'y' is already present. Adding another scale for 'y', which will\n",
      "replace the existing scale.\n",
      "\n",
      "Scale for 'y' is already present. Adding another scale for 'y', which will\n",
      "replace the existing scale.\n",
      "\n",
      "Scale for 'y' is already present. Adding another scale for 'y', which will\n",
      "replace the existing scale.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "###Figure 3C, Violin plot for marker genes\n",
    "\n",
    "dir.name <- \"3C\"\n",
    "file.name <- \"Fig3C.Vlnplot.Cluster.tiff\"\n",
    "genes <- c(\"Mki67\", \"Isg15\",\n",
    "           \"Pdcd1\", \"Havcr2\", \n",
    "           \"Ifng\", \"Gzmb\", \"Tnfrsf9\", \n",
    "           \"Tcf7\", \"Il7r\", \"Sell\")\n",
    "\n",
    "################################# Processing layer ##########################################\n",
    "dir.create(dir.name)\n",
    "\n",
    "ppi <- 600\n",
    "name.output <- str_c(dir.name, file.name, sep = \"/\")\n",
    "tiff(name.output, width=2*ppi, height=2.4*ppi, res=ppi)\n",
    "p <- StackedVlnPlot(obj = seu, features = genes)\n",
    "plot(p)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "36c6f8b9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(dir.name):\n",
      "“ '3D' はすでに存在します ”\n"
     ]
    }
   ],
   "source": [
    "#Figure3D, Plot gene scores\n",
    "\n",
    "dir.input <- \"Signatures\"\n",
    "dir.name <- \"3D\"\n",
    "\n",
    "################################# Processing layer ##########################################\n",
    "dir.create(dir.name)\n",
    "files  <- list.files(dir.input, pattern=\"Sig.\")\n",
    "\n",
    "#Violin Plot\n",
    "for(i in files){\n",
    "  sig_name <- str_remove(i, \"Sig.\")\n",
    "  sig_name <- str_remove(sig_name, \".txt\")\n",
    "  feature <- paste(sig_name, \"1\", sep = \"\")\n",
    "  \n",
    "  file.name2 <- paste(sig_name, \"tiff\", sep = \".\")\n",
    "  name.output <- str_c(dir.name, file.name2, sep = \"/\")\n",
    "  ppi <- 600\n",
    "  tiff(name.output, width=1.5*ppi, height=1*ppi, res=ppi)\n",
    "  p <- modify_vlnplot_score(obj = seu, feature = feature, pt.size = 0)\n",
    "  plot(p)\n",
    "  dev.off()\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "c8dca765",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(dir.name):\n",
      "“ '3E' はすでに存在します ”\n"
     ]
    }
   ],
   "source": [
    "#Figure3E: Plot clones of Oligoclonal / Polyclonal fraction in Dimplot\n",
    "\n",
    "modules <- c(\"Oligo\", \"Poly\", \"nonOL\")\n",
    "dir.name <- \"3E\"\n",
    "#Specify clone table merged with dLN overlap status\n",
    "name.clone.table <- \"clone_within_cluster.dLNOL.csv\"\n",
    "\n",
    "############################# Processing layer ######################################\n",
    "dir.create(dir.name)\n",
    "\n",
    "#Call clone table merged with dLN overlap status (Stored in \"scTCR_WGCNA_combine\" directory)\n",
    "name.clone.table <- str_c(\"scTCR_analysis\", name.clone.table, sep = \"/\")\n",
    "clone_within_cluster <- read.csv(name.clone.table, header = TRUE)\n",
    "\n",
    "#Call metadata and extract cell names\n",
    "meta.data <- as.data.frame(seu@meta.data)\n",
    "tmp_raw <- as.vector(row.names(meta.data))\n",
    "\n",
    "#Output twice size of figure, compress after pasting in the powerpoint.\n",
    "for(i in modules){\n",
    "    #Extract clones in each subset\n",
    "    clone_plot <- dplyr::filter(clone_within_cluster, OL == i)\n",
    "    clone_names <- clone_plot$names\n",
    "\n",
    "    #Extract cell names for Dimplot\n",
    "    data_sub <- dplyr::filter(meta.data, clone.id %in% clone_names)\n",
    "    clone.id <- as.vector(row.names(data_sub))\n",
    "    tmp2 = tmp_raw %in% clone.id\n",
    "    \n",
    "    #Output\n",
    "    name.output <- str_c(i, \"clone.plot.tiff\", sep = \".\")\n",
    "    name.output <- str_c(dir.name, name.output, sep = \"/\")\n",
    "    PlotClones(seu, tmp2, name.output, meta.data, clone_plot, \"umap\")\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "639d2cf2",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(dir.name):\n",
      "“ '3G' はすでに存在します ”\n"
     ]
    }
   ],
   "source": [
    "#Figure 3G: Correlation analysis between gene score and clone size\n",
    "#Input; scTCR_analysis/clone_within_cluster.dLNOL.csv\n",
    "\n",
    "file.name  <- \"clone_within_cluster.dLNOL.csv\"\n",
    "dir.input <- \"scTCR_analysis\"\n",
    "dir.name <- \"3G\"\n",
    "\n",
    "############################# Processing layer ######################################\n",
    "dir.create(dir.name)\n",
    "\n",
    "name.input <- str_c(dir.input, file.name, sep = \"/\")\n",
    "d <- read.csv(name.input, header = TRUE)\n",
    "d_sub <- dplyr::select(d, c(\"temp_freq\", \"Cytotoxicity1\", \"Proliferation1\", \"Tumor.Prog1\", \"Tumor.Term1\", \"OL\"))\n",
    "d_sub <- dplyr::filter(d_sub, temp_freq > 0.001)\n",
    "d_sub <- dplyr::filter(d_sub, OL %in% c(\"Poly\", \"Oligo\"))\n",
    "\n",
    "axis_x <- \"temp_freq\"\n",
    "axis_y <- c(\"Tumor.Prog1\", \"Tumor.Term1\")\n",
    "\n",
    "#Convert frequency to percentage\n",
    "d_sub$temp_freq <- d_sub$temp_freq*100\n",
    "\n",
    "for(i in axis_x){\n",
    "  for(j in axis_y){\n",
    "\n",
    "    #Extract data\n",
    "    d_sub2 <- dplyr::select(d_sub, c(\"OL\", i, j))\n",
    "    names(d_sub2) <- c(\"OL\", \"x_axis\", \"y_axis\")\n",
    "    \n",
    "    #Output\n",
    "    ppi <- 600\n",
    "    name_out <- str_c(\"x\", i, \"y\", j, \"tiff\", sep = \".\")\n",
    "    name_out <- str_c(dir.name, name_out, sep = \"/\")\n",
    "    tiff(name_out, width=1.5*ppi, height=1.5*ppi, res=ppi)\n",
    "    p <- ggplot(d_sub2, aes(x=x_axis, y=y_axis, colour=OL)) +\n",
    "      scale_colour_manual(values=c(Poly=\"blue\", Oligo=\"red\"))+\n",
    "      geom_point(size=0.8) +\n",
    "      scale_x_log10() +\n",
    "      theme_bw(base_size = 6) +\n",
    "      theme(\n",
    "        axis.title.x=element_blank(), axis.title.y=element_blank(),\n",
    "        axis.text.x = element_text(size=6, family=\"Arial\"),\n",
    "        axis.text.y = element_text(size=6, family=\"Arial\")) +\n",
    "      guides(colour=FALSE)\n",
    "    plot(p)\n",
    "    dev.off()\n",
    "      \n",
    "    #Calculate correlation coefficient\n",
    "    result.name <- str_c(dir.name, \"Genescore.correlation.txt\", sep = \"/\")\n",
    "    cor_t <- cor.test(log10(d_sub2$x_axis), d_sub2$y_axis, use=\"pairwise\", method=\"pearson\")\n",
    "    pvalue <- cor_t$p.value\n",
    "    rvalue <- cor_t$estimate\n",
    "    cor_abs <- abs(rvalue)\n",
    "    \n",
    "    #Record pvalues\n",
    "    tail <- \"tumor repertoire rvalue r_abs pvalue\\n\"\n",
    "    cor <- sprintf(\"%f\\t\", rvalue)\n",
    "    cor_abs <- sprintf(\"%f\\t\", cor_abs)\n",
    "    p <- sprintf(\"%f\\t\", pvalue)\n",
    "    freq3 <- paste(i, j, cor, cor_abs, p, tail)\n",
    "    cat(freq3, file = result.name, append = TRUE)\n",
    "  }\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a8a5512",
   "metadata": {},
   "source": [
    "### Figure S2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "50da054f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(dir.name):\n",
      "“ 'S2A' はすでに存在します ”\n"
     ]
    }
   ],
   "source": [
    "###Figure S2A: Marker Gene Heatmap\n",
    "\n",
    "#Specify Merker gene list\n",
    "dir.input <- \"MarkerGenePlot\"\n",
    "file.name <- \"Aoki_3rdALLmarkers_minpct0.1_Adj_p0.05.txt\"\n",
    "\n",
    "top_count <- 5 #Number of genes per cluster\n",
    "dir.name <- \"S2A\"\n",
    "\n",
    "############################# Processing layer ######################################\n",
    "dir.create(dir.name)\n",
    "\n",
    "name.input <- str_c(dir.input, file.name, sep = \"/\")\n",
    "seu.markers = read.table(name.input, header = TRUE)\n",
    "\n",
    "#Extract genes for plotting\n",
    "top10 = seu.markers %>% group_by(cluster) %>% top_n(top_count, avg_log2FC)\n",
    "top10 = as.data.frame(top10)\n",
    "top10  = top10 [!duplicated(top10$gene),]\n",
    "top10 = top10 %>% arrange(desc(avg_log2FC))  %>% arrange(cluster)\n",
    "top10 = as.data.frame(top10)\n",
    "\n",
    "#Output heatmap\n",
    "file.name <- paste(dir.name, \"S2A.MarkerHeatmap.tiff\", sep = \"/\")\n",
    "p <- DoHeatmap(seu, features = top10$gene, disp.min = -2.5, disp.max = 2.5,\n",
    "               size = 3, angle = 0) +\n",
    "        theme(legend.position = 'bottom',\n",
    "             axis.text.y = element_text(size=6, family=\"Arial\"),\n",
    "             legend.title = element_text(size=8, family=\"Arial\"),\n",
    "             legend.key.height = unit(0.33, 'cm'),\n",
    "             legend.text = element_text(size=6, family=\"Arial\")) +\n",
    "        guides(colour=FALSE)\n",
    "ggsave(file = file.name, plot = p, device=\"tiff\", units=\"in\", dpi = 300,\n",
    "       width = 3.3, height = 5.25, limitsize=FALSE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "a80fd89a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(dir.name):\n",
      "“ 'S2B' はすでに存在します ”\n"
     ]
    }
   ],
   "source": [
    "###Figure S2B: Scatter plot for marker genes\n",
    "#Output four times size of figure, compress after pasting in the powerpoint.\n",
    "\n",
    "tmp = c(\"Pdcd1\", \"Havcr2\", \"Tigit\", \"Tnfrsf9\", \n",
    "            \"Ifng\", \"Gzmb\", \"Klrg1\", \"Cd69\",\n",
    "            \"Ccr7\", \"Tcf7\", \"Il7r\", \"Sell\", \n",
    "            \"Cd38\", \"Mki67\", \"Mx1\", \"Isg15\")\n",
    "dir.name <- \"S2B\"\n",
    "\n",
    "############################# Processing layer ######################################\n",
    "dir.create(dir.name)\n",
    "\n",
    "file.name = paste(dir.name, \"Marker.Expression.tiff\", sep='/')\n",
    "ppi <- 600\n",
    "p <- FeaturePlot(seu, features = tmp, ncol = 4, order = TRUE,\n",
    "                reduction = \"umap\", dims=c(1,2), cols = c(\"grey\", \"red\"), pt.size = 0.1) *\n",
    "theme(legend.position = 'none',\n",
    "      axis.title.x = element_blank(), \n",
    "        axis.title.y = element_blank(),\n",
    "        axis.text.x = element_blank(), \n",
    "        axis.text.y = element_blank(),\n",
    "        plot.title= element_text(size=24, family = \"Arial\"))\n",
    "save_plot(file = file.name, plot(p), device=\"tiff\", \n",
    "          units=\"in\", dpi = 150, base_width = 11.2, base_height = 12.4, limitsize=FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "7c88470d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(dir.name):\n",
      "“'S2D' already exists”\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "###Figure S2D: Pseudotime Analysis between clusters\n",
    "\n",
    "dir.name <- \"S2D\"\n",
    "start <- \"7\"\n",
    "seu.sub <- seu\n",
    "name <- \"pseudotime.all\"\n",
    "red.use <- \"umap\"\n",
    "\n",
    "############################# Processing layer ######################################\n",
    "dir.create(dir.name)\n",
    "\n",
    "#Define PCs for Pseudotime analysis (Call the result of Jackstraw)\n",
    "tmp = as.data.frame(seu@reductions$pca@jackstraw@overall.p.values)\n",
    "tmp1 = tmp[tmp$Score>0.05,1]\n",
    "dims= c(1:(min(tmp1)-1))\n",
    "\n",
    "#Perform pseudotime analysis\n",
    "seu2 <- Pseudotime.Min(seu, seu.sub, name, dims, start, dir.name)\n",
    "\n",
    "#Plot pseudotime in dimentional reduction\n",
    "name.output=str_c(dir.name, name, sep='/') %>% str_c(\"from\", start, red.use, \"pseudotime.tiff\", sep = \".\")\n",
    "ppi <- 600\n",
    "tiff(name.output, width=3.8*ppi, height=3*ppi, res=ppi)\n",
    "#Output twice size of figure, compress after pasting in the powerpoint.\n",
    "FeaturePlot(object = seu2, features = name, cols = c(\"#800080\", \"#ffee00\"),\n",
    "            reduction = red.use, pt.size = 0.1) +\n",
    "  theme(plot.title = element_blank(), \n",
    "        axis.title.x = element_text(size=16, family = \"Arial\"), \n",
    "        axis.title.y = element_text(size=16, family = \"Arial\"), \n",
    "        axis.text.x = element_text(size=12, colour = 1, family = \"Arial\"), \n",
    "        axis.text.y = element_text(size = 12, colour = 1, family = \"Arial\"),\n",
    "       legend.title = element_text(size=12, family = \"Arial\"),\n",
    "       legend.text = element_text(size=12, family = \"Arial\"),\n",
    "       legend.key.width = unit(0.5, \"cm\")) +\n",
    "  theme(panel.border = element_rect(fill = NA, size = 0.75)) \n",
    "dev.off()\n",
    "    \n",
    "#Plot Pseudotime in violin plot\n",
    "name.output=str_c(dir.name, name, sep='/') %>% str_c(\"from\", start, \"violin\", \"pseudotime.tiff\", sep = \".\") \n",
    "ppi <- 600\n",
    "tiff(name.output, width=1.5*ppi, height=1*ppi, res=ppi)\n",
    "p <- modify_vlnplot_score(obj = seu2, feature = name, pt.size = 0)\n",
    "plot(p)\n",
    "dev.off()\n",
    "\n",
    "#save seurat object with pseudotime\n",
    "file.name=str_replace(name.input, \".rda\", \".pseudotime.rda\")\n",
    "#save(seu2, file=file.name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "id": "1ecd600d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(dir.name):\n",
      "“ 'scTCR_analysis' はすでに存在します ”\n"
     ]
    }
   ],
   "source": [
    "###Figure S3E: Calculate clonality of TCR repertoire per clusters\n",
    "\n",
    "#Specify clone table merged with dLN overlap status\n",
    "file.name  <- \"clone_within_cluster.dLNOL.csv\"\n",
    "dir.name <- \"scTCR_analysis\"\n",
    "\n",
    "############################# Processing layer ######################################\n",
    "dir.create(dir.name)\n",
    "name.input <- str_c(dir.name, file.name, sep = \"/\")\n",
    "d <- read.csv(name.input, header = TRUE)\n",
    "\n",
    "clust.array <- c(\"X0\", \"X1\", \"X2\", \"X3\", \"X4\", \"X5\", \"X6\", \"X7\")\n",
    "\n",
    "output <- data.frame()\n",
    "\n",
    "for(i in 1:length(clust.array)){\n",
    "  clust <- clust.array[i]\n",
    "  d_sub <- select(d, clust)\n",
    "  names(d_sub) <- \"count\"\n",
    "  d_sub <- filter(d_sub, count > 0)\n",
    "  d_sub$freq <- d_sub$count / sum(d_sub$count)\n",
    "  \n",
    "  ##Calculate median and count clonotypes\n",
    "  \n",
    "  l <- length(d_sub$freq)\n",
    "  count <- sum(d_sub$count)\n",
    "  \n",
    "  #Calculate 1-Pielou clonality\n",
    "  e <- log(d_sub$freq)\n",
    "  pi <- 1+sum(d_sub$freq * e)/log(length(d_sub$freq))\n",
    "  \n",
    "  out <- c(l, count, pi)\n",
    "  \n",
    "  output <- rbind(output, out)\n",
    "}\n",
    "\n",
    "names(output) <- c(\"clones\", \"cells\", \"Pielou\")\n",
    "output$cluster <- clust.array\n",
    "\n",
    "name.output <- str_c(dir.name, \"S2D.Clonality_inCluster.csv\", sep = \"/\")\n",
    "\n",
    "write.csv(output, name.output, row.names = F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "ff46b2ed",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(dir.name):\n",
      "“ 'S2E' はすでに存在します ”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Formal class 'Seurat' [package \"SeuratObject\"] with 13 slots\n",
      "  ..@ assays      :List of 1\n",
      "  .. ..$ RNA:Formal class 'Assay' [package \"SeuratObject\"] with 8 slots\n",
      "  .. .. .. ..@ counts       :Formal class 'dgCMatrix' [package \"Matrix\"] with 6 slots\n",
      "  .. .. .. .. .. ..@ i       : int [1:878484] 9 12 14 15 16 17 20 26 28 33 ...\n",
      "  .. .. .. .. .. ..@ p       : int [1:9114] 0 106 196 279 338 410 491 593 664 769 ...\n",
      "  .. .. .. .. .. ..@ Dim     : int [1:2] 371 9113\n",
      "  .. .. .. .. .. ..@ Dimnames:List of 2\n",
      "  .. .. .. .. .. .. ..$ : chr [1:371] \"Ada\" \"Adgre1\" \"Adgrg3\" \"Anxa5\" ...\n",
      "  .. .. .. .. .. .. ..$ : chr [1:9113] \"mouseSampleTag5_Aoki_5\" \"mouseSampleTag6_Aoki_21\" \"mouseSampleTag6_Aoki_49\" \"mouseSampleTag5_Aoki_62\" ...\n",
      "  .. .. .. .. .. ..@ x       : num [1:878484] 1 94 36 28 51 247 920 109 1 9 ...\n",
      "  .. .. .. .. .. ..@ factors : list()\n",
      "  .. .. .. ..@ data         :Formal class 'dgCMatrix' [package \"Matrix\"] with 6 slots\n",
      "  .. .. .. .. .. ..@ i       : int [1:878484] 9 12 14 15 16 17 20 26 28 33 ...\n",
      "  .. .. .. .. .. ..@ p       : int [1:9114] 0 106 196 279 338 410 491 593 664 769 ...\n",
      "  .. .. .. .. .. ..@ Dim     : int [1:2] 371 9113\n",
      "  .. .. .. .. .. ..@ Dimnames:List of 2\n",
      "  .. .. .. .. .. .. ..$ : chr [1:371] \"Ada\" \"Adgre1\" \"Adgrg3\" \"Anxa5\" ...\n",
      "  .. .. .. .. .. .. ..$ : chr [1:9113] \"mouseSampleTag5_Aoki_5\" \"mouseSampleTag6_Aoki_21\" \"mouseSampleTag6_Aoki_49\" \"mouseSampleTag5_Aoki_62\" ...\n",
      "  .. .. .. .. .. ..@ x       : num [1:878484] 3.59 8.1 7.14 6.89 7.49 ...\n",
      "  .. .. .. .. .. ..@ factors : list()\n",
      "  .. .. .. ..@ scale.data   : num [1:371, 1:9113] -0.2928 -0.0614 -0.1665 -1.0406 -0.0973 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:371] \"Ada\" \"Adgre1\" \"Adgrg3\" \"Anxa5\" ...\n",
      "  .. .. .. .. .. ..$ : chr [1:9113] \"mouseSampleTag5_Aoki_5\" \"mouseSampleTag6_Aoki_21\" \"mouseSampleTag6_Aoki_49\" \"mouseSampleTag5_Aoki_62\" ...\n",
      "  .. .. .. ..@ key          : chr \"rna_\"\n",
      "  .. .. .. ..@ assay.orig   : NULL\n",
      "  .. .. .. ..@ var.features : chr [1:371] \"Ada\" \"Adgre1\" \"Adgrg3\" \"Anxa5\" ...\n",
      "  .. .. .. ..@ meta.features:'data.frame':\t371 obs. of  0 variables\n",
      "  .. .. .. ..@ misc         : list()\n",
      "  ..@ meta.data   :'data.frame':\t9113 obs. of  8 variables:\n",
      "  .. ..$ orig.ident     : Factor w/ 7 levels \"doublet\",\"mouseSampleTag5\",..: 2 3 3 2 3 3 3 2 3 2 ...\n",
      "  .. ..$ nCount_RNA     : num [1:9113] 28443 17244 24761 5404 12108 ...\n",
      "  .. ..$ nFeature_RNA   : int [1:9113] 106 90 83 59 72 81 102 71 105 109 ...\n",
      "  .. ..$ RNA_snn_res.1  : Factor w/ 14 levels \"0\",\"1\",\"2\",\"3\",..: 3 2 2 4 4 3 6 4 6 6 ...\n",
      "  .. ..$ seurat_clusters: Factor w/ 8 levels \"0\",\"1\",\"2\",\"3\",..: 3 3 2 4 4 3 1 4 2 1 ...\n",
      "  .. ..$ RNA_snn_res.0.6: Factor w/ 8 levels \"0\",\"1\",\"2\",\"3\",..: 3 3 2 4 4 3 1 4 2 1 ...\n",
      "  .. ..$ clone.id       : chr [1:9113] \"TGTGCTTTGGATGATAACTATGCCCAGGGATTAACCTTC_TRAV13-1_TRAJ26_TGCACCTGCAGTGCTCTGGGGGGGCGGGCTGAGCAGTTCTTC_TRBV1_TRBJ2-1\" \"TGTGCTCTGGGTCATGACACAAATGCTTACAAAGTCATCTTT_TRAV6-7-DV9_TRAJ3_TGTGCCAGCAGTGATCCGGGACAGAACCAGGCTCCGCTTTTT_TRBV13-3_TRBJ1-5\" \"TGTGCTCTGGGTCATGACACAAATGCTTACAAAGTCATCTTT_TRAV6-7-DV9_TRAJ3_TGTGCCAGCAGTGATCCGGGACAGAACCAGGCTCCGCTTTTT_TRBV13-3_TRBJ1-5\" \"TGTGCTACAGGTCCGGACTACAGCAACAACAGACTTACTTTG_TRAV8D-2_TRAJ7_TGCACCTGCAGTGCTCTGGGGGGGCGGGCTGAGCAGTTCTTC_TRBV1_TRBJ2-1\" ...\n",
      "  .. ..$ active.ident   : Factor w/ 2 levels \"Oligo\",\"Poly\": 1 1 1 2 1 1 2 1 1 1 ...\n",
      "  ..@ active.assay: chr \"RNA\"\n",
      "  ..@ active.ident: Factor w/ 2 levels \"Oligo\",\"Poly\": 1 1 1 2 1 1 2 1 1 1 ...\n",
      "  .. ..- attr(*, \"names\")= chr [1:9113] \"mouseSampleTag5_Aoki_5\" \"mouseSampleTag6_Aoki_21\" \"mouseSampleTag6_Aoki_49\" \"mouseSampleTag5_Aoki_62\" ...\n",
      "  ..@ graphs      : list()\n",
      "  ..@ neighbors   : list()\n",
      "  ..@ reductions  :List of 3\n",
      "  .. ..$ pca :Formal class 'DimReduc' [package \"SeuratObject\"] with 9 slots\n",
      "  .. .. .. ..@ cell.embeddings           : num [1:9113, 1:100] -0.517 -0.636 -1.461 0.914 -1.753 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:9113] \"mouseSampleTag5_Aoki_5\" \"mouseSampleTag6_Aoki_21\" \"mouseSampleTag6_Aoki_49\" \"mouseSampleTag5_Aoki_62\" ...\n",
      "  .. .. .. .. .. ..$ : chr [1:100] \"PC_1\" \"PC_2\" \"PC_3\" \"PC_4\" ...\n",
      "  .. .. .. ..@ feature.loadings          : num [1:369, 1:100] 0.0168 0.00095 -0.01619 0.00564 0.00488 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:369] \"Ada\" \"Adgre1\" \"Adgrg3\" \"Anxa5\" ...\n",
      "  .. .. .. .. .. ..$ : chr [1:100] \"PC_1\" \"PC_2\" \"PC_3\" \"PC_4\" ...\n",
      "  .. .. .. ..@ feature.loadings.projected: num[0 , 0 ] \n",
      "  .. .. .. ..@ assay.used                : chr \"RNA\"\n",
      "  .. .. .. ..@ global                    : logi FALSE\n",
      "  .. .. .. ..@ stdev                     : num [1:100] 2.76 2.27 2.04 1.92 1.8 ...\n",
      "  .. .. .. ..@ key                       : chr \"PC_\"\n",
      "  .. .. .. ..@ jackstraw                 :Formal class 'JackStrawData' [package \"SeuratObject\"] with 4 slots\n",
      "  .. .. .. .. .. ..@ empirical.p.values     : num[0 , 0 ] \n",
      "  .. .. .. .. .. ..@ fake.reduction.scores  : num[0 , 0 ] \n",
      "  .. .. .. .. .. ..@ empirical.p.values.full: num[0 , 0 ] \n",
      "  .. .. .. .. .. ..@ overall.p.values       : num[0 , 0 ] \n",
      "  .. .. .. ..@ misc                      :List of 1\n",
      "  .. .. .. .. ..$ total.variance: num 288\n",
      "  .. ..$ tsne:Formal class 'DimReduc' [package \"SeuratObject\"] with 9 slots\n",
      "  .. .. .. ..@ cell.embeddings           : num [1:9113, 1:2] -21.59 -17.92 26.7 -4.27 20.02 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:9113] \"mouseSampleTag5_Aoki_5\" \"mouseSampleTag6_Aoki_21\" \"mouseSampleTag6_Aoki_49\" \"mouseSampleTag5_Aoki_62\" ...\n",
      "  .. .. .. .. .. ..$ : chr [1:2] \"fitSNE_1\" \"fitSNE_2\"\n",
      "  .. .. .. ..@ feature.loadings          : num[0 , 0 ] \n",
      "  .. .. .. ..@ feature.loadings.projected: num[0 , 0 ] \n",
      "  .. .. .. ..@ assay.used                : chr \"RNA\"\n",
      "  .. .. .. ..@ global                    : logi TRUE\n",
      "  .. .. .. ..@ stdev                     : num(0) \n",
      "  .. .. .. ..@ key                       : chr \"fitSNE_\"\n",
      "  .. .. .. ..@ jackstraw                 :Formal class 'JackStrawData' [package \"SeuratObject\"] with 4 slots\n",
      "  .. .. .. .. .. ..@ empirical.p.values     : num[0 , 0 ] \n",
      "  .. .. .. .. .. ..@ fake.reduction.scores  : num[0 , 0 ] \n",
      "  .. .. .. .. .. ..@ empirical.p.values.full: num[0 , 0 ] \n",
      "  .. .. .. .. .. ..@ overall.p.values       : num[0 , 0 ] \n",
      "  .. .. .. ..@ misc                      : list()\n",
      "  .. ..$ umap:Formal class 'DimReduc' [package \"SeuratObject\"] with 9 slots\n",
      "  .. .. .. ..@ cell.embeddings           : num [1:9113, 1:2] -2.82 -2.48 -2.05 -3.47 -1.95 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:9113] \"mouseSampleTag5_Aoki_5\" \"mouseSampleTag6_Aoki_21\" \"mouseSampleTag6_Aoki_49\" \"mouseSampleTag5_Aoki_62\" ...\n",
      "  .. .. .. .. .. ..$ : chr [1:2] \"UMAP_1\" \"UMAP_2\"\n",
      "  .. .. .. ..@ feature.loadings          : num[0 , 0 ] \n",
      "  .. .. .. ..@ feature.loadings.projected: num[0 , 0 ] \n",
      "  .. .. .. ..@ assay.used                : chr \"RNA\"\n",
      "  .. .. .. ..@ global                    : logi TRUE\n",
      "  .. .. .. ..@ stdev                     : num(0) \n",
      "  .. .. .. ..@ key                       : chr \"UMAP_\"\n",
      "  .. .. .. ..@ jackstraw                 :Formal class 'JackStrawData' [package \"SeuratObject\"] with 4 slots\n",
      "  .. .. .. .. .. ..@ empirical.p.values     : num[0 , 0 ] \n",
      "  .. .. .. .. .. ..@ fake.reduction.scores  : num[0 , 0 ] \n",
      "  .. .. .. .. .. ..@ empirical.p.values.full: num[0 , 0 ] \n",
      "  .. .. .. .. .. ..@ overall.p.values       : num[0 , 0 ] \n",
      "  .. .. .. ..@ misc                      : list()\n",
      "  ..@ images      : list()\n",
      "  ..@ project.name: chr \"seu\"\n",
      "  ..@ misc        : list()\n",
      "  ..@ version     :Classes 'package_version', 'numeric_version'  hidden list of 1\n",
      "  .. ..$ : int [1:3] 4 0 0\n",
      "  ..@ commands    :List of 9\n",
      "  .. ..$ NormalizeData.RNA    :Formal class 'SeuratCommand' [package \"SeuratObject\"] with 5 slots\n",
      "  .. .. .. ..@ name       : chr \"NormalizeData.RNA\"\n",
      "  .. .. .. ..@ time.stamp : POSIXct[1:1], format: \"2022-02-10 08:54:59\"\n",
      "  .. .. .. ..@ assay.used : chr \"RNA\"\n",
      "  .. .. .. ..@ call.string: chr \"NormalizeData(object = seu, scale.factor = 1e+06)\"\n",
      "  .. .. .. ..@ params     :List of 5\n",
      "  .. .. .. .. ..$ assay               : chr \"RNA\"\n",
      "  .. .. .. .. ..$ normalization.method: chr \"LogNormalize\"\n",
      "  .. .. .. .. ..$ scale.factor        : num 1e+06\n",
      "  .. .. .. .. ..$ margin              : num 1\n",
      "  .. .. .. .. ..$ verbose             : logi TRUE\n",
      "  .. ..$ ScaleData.RNA        :Formal class 'SeuratCommand' [package \"SeuratObject\"] with 5 slots\n",
      "  .. .. .. ..@ name       : chr \"ScaleData.RNA\"\n",
      "  .. .. .. ..@ time.stamp : POSIXct[1:1], format: \"2022-02-10 08:55:03\"\n",
      "  .. .. .. ..@ assay.used : chr \"RNA\"\n",
      "  .. .. .. ..@ call.string: chr [1:2] \"ScaleData(object = seu, vars.to.regress = c(\\\"nCount_RNA\\\"), \" \"    features = all.genes)\"\n",
      "  .. .. .. ..@ params     :List of 11\n",
      "  .. .. .. .. ..$ features          : chr [1:371] \"Ada\" \"Adgre1\" \"Adgrg3\" \"Anxa5\" ...\n",
      "  .. .. .. .. ..$ assay             : chr \"RNA\"\n",
      "  .. .. .. .. ..$ vars.to.regress   : chr \"nCount_RNA\"\n",
      "  .. .. .. .. ..$ model.use         : chr \"linear\"\n",
      "  .. .. .. .. ..$ use.umi           : logi FALSE\n",
      "  .. .. .. .. ..$ do.scale          : logi TRUE\n",
      "  .. .. .. .. ..$ do.center         : logi TRUE\n",
      "  .. .. .. .. ..$ scale.max         : num 10\n",
      "  .. .. .. .. ..$ block.size        : num 1000\n",
      "  .. .. .. .. ..$ min.cells.to.block: num 3000\n",
      "  .. .. .. .. ..$ verbose           : logi TRUE\n",
      "  .. ..$ RunPCA.RNA           :Formal class 'SeuratCommand' [package \"SeuratObject\"] with 5 slots\n",
      "  .. .. .. ..@ name       : chr \"RunPCA.RNA\"\n",
      "  .. .. .. ..@ time.stamp : POSIXct[1:1], format: \"2022-02-10 08:55:10\"\n",
      "  .. .. .. ..@ assay.used : chr \"RNA\"\n",
      "  .. .. .. ..@ call.string: chr [1:2] \"RunPCA(object = seu, features = seu@assays$RNA@var.features, \" \"    npcs = 100)\"\n",
      "  .. .. .. ..@ params     :List of 11\n",
      "  .. .. .. .. ..$ assay          : chr \"RNA\"\n",
      "  .. .. .. .. ..$ features       : chr [1:371] \"Ada\" \"Adgre1\" \"Adgrg3\" \"Anxa5\" ...\n",
      "  .. .. .. .. ..$ npcs           : num 100\n",
      "  .. .. .. .. ..$ rev.pca        : logi FALSE\n",
      "  .. .. .. .. ..$ weight.by.var  : logi TRUE\n",
      "  .. .. .. .. ..$ verbose        : logi TRUE\n",
      "  .. .. .. .. ..$ ndims.print    : int [1:5] 1 2 3 4 5\n",
      "  .. .. .. .. ..$ nfeatures.print: num 30\n",
      "  .. .. .. .. ..$ reduction.name : chr \"pca\"\n",
      "  .. .. .. .. ..$ reduction.key  : chr \"PC_\"\n",
      "  .. .. .. .. ..$ seed.use       : num 42\n",
      "  .. ..$ JackStraw.RNA.pca    :Formal class 'SeuratCommand' [package \"SeuratObject\"] with 5 slots\n",
      "  .. .. .. ..@ name       : chr \"JackStraw.RNA.pca\"\n",
      "  .. .. .. ..@ time.stamp : POSIXct[1:1], format: \"2022-02-10 09:07:34\"\n",
      "  .. .. .. ..@ assay.used : chr \"RNA\"\n",
      "  .. .. .. ..@ call.string: chr \"JackStraw(object = seu, num.replicate = 100, dims = 100)\"\n",
      "  .. .. .. ..@ params     :List of 7\n",
      "  .. .. .. .. ..$ reduction    : chr \"pca\"\n",
      "  .. .. .. .. ..$ assay        : chr \"RNA\"\n",
      "  .. .. .. .. ..$ dims         : num 100\n",
      "  .. .. .. .. ..$ num.replicate: num 100\n",
      "  .. .. .. .. ..$ prop.freq    : num 0.01\n",
      "  .. .. .. .. ..$ verbose      : logi TRUE\n",
      "  .. .. .. .. ..$ maxit        : num 1000\n",
      "  .. ..$ ScoreJackStraw       :Formal class 'SeuratCommand' [package \"SeuratObject\"] with 5 slots\n",
      "  .. .. .. ..@ name       : chr \"ScoreJackStraw\"\n",
      "  .. .. .. ..@ time.stamp : POSIXct[1:1], format: \"2022-02-10 09:07:35\"\n",
      "  .. .. .. ..@ assay.used : chr \"RNA\"\n",
      "  .. .. .. ..@ call.string: chr \"ScoreJackStraw(object = seu, dims = 1:100, score.thresh = 0.05)\"\n",
      "  .. .. .. ..@ params     :List of 4\n",
      "  .. .. .. .. ..$ reduction   : chr \"pca\"\n",
      "  .. .. .. .. ..$ dims        : int [1:100] 1 2 3 4 5 6 7 8 9 10 ...\n",
      "  .. .. .. .. ..$ score.thresh: num 0.05\n",
      "  .. .. .. .. ..$ do.plot     : logi FALSE\n",
      "  .. ..$ FindNeighbors.RNA.pca:Formal class 'SeuratCommand' [package \"SeuratObject\"] with 5 slots\n",
      "  .. .. .. ..@ name       : chr \"FindNeighbors.RNA.pca\"\n",
      "  .. .. .. ..@ time.stamp : POSIXct[1:1], format: \"2022-02-10 09:08:12\"\n",
      "  .. .. .. ..@ assay.used : chr \"RNA\"\n",
      "  .. .. .. ..@ call.string: chr \"FindNeighbors(seu, reduction = \\\"pca\\\", dims = dims, force.recalc = TRUE)\"\n",
      "  .. .. .. ..@ params     :List of 17\n",
      "  .. .. .. .. ..$ reduction      : chr \"pca\"\n",
      "  .. .. .. .. ..$ dims           : int [1:18] 1 2 3 4 5 6 7 8 9 10 ...\n",
      "  .. .. .. .. ..$ assay          : chr \"RNA\"\n",
      "  .. .. .. .. ..$ k.param        : num 20\n",
      "  .. .. .. .. ..$ return.neighbor: logi FALSE\n",
      "  .. .. .. .. ..$ compute.SNN    : logi TRUE\n",
      "  .. .. .. .. ..$ prune.SNN      : num 0.0667\n",
      "  .. .. .. .. ..$ nn.method      : chr \"annoy\"\n",
      "  .. .. .. .. ..$ n.trees        : num 50\n",
      "  .. .. .. .. ..$ annoy.metric   : chr \"euclidean\"\n",
      "  .. .. .. .. ..$ nn.eps         : num 0\n",
      "  .. .. .. .. ..$ verbose        : logi TRUE\n",
      "  .. .. .. .. ..$ force.recalc   : logi TRUE\n",
      "  .. .. .. .. ..$ do.plot        : logi FALSE\n",
      "  .. .. .. .. ..$ graph.name     : chr [1:2] \"RNA_nn\" \"RNA_snn\"\n",
      "  .. .. .. .. ..$ l2.norm        : logi FALSE\n",
      "  .. .. .. .. ..$ cache.index    : logi FALSE\n",
      "  .. ..$ FindClusters         :Formal class 'SeuratCommand' [package \"SeuratObject\"] with 5 slots\n",
      "  .. .. .. ..@ name       : chr \"FindClusters\"\n",
      "  .. .. .. ..@ time.stamp : POSIXct[1:1], format: \"2022-02-10 09:08:15\"\n",
      "  .. .. .. ..@ assay.used : chr \"RNA\"\n",
      "  .. .. .. ..@ call.string: chr \"FindClusters(object = seu, resolution = resol)\"\n",
      "  .. .. .. ..@ params     :List of 10\n",
      "  .. .. .. .. ..$ graph.name      : chr \"RNA_snn\"\n",
      "  .. .. .. .. ..$ modularity.fxn  : num 1\n",
      "  .. .. .. .. ..$ resolution      : num 0.6\n",
      "  .. .. .. .. ..$ method          : chr \"matrix\"\n",
      "  .. .. .. .. ..$ algorithm       : num 1\n",
      "  .. .. .. .. ..$ n.start         : num 10\n",
      "  .. .. .. .. ..$ n.iter          : num 10\n",
      "  .. .. .. .. ..$ random.seed     : num 0\n",
      "  .. .. .. .. ..$ group.singletons: logi TRUE\n",
      "  .. .. .. .. ..$ verbose         : logi TRUE\n",
      "  .. ..$ RunTSNE              :Formal class 'SeuratCommand' [package \"SeuratObject\"] with 5 slots\n",
      "  .. .. .. ..@ name       : chr \"RunTSNE\"\n",
      "  .. .. .. ..@ time.stamp : POSIXct[1:1], format: \"2022-02-10 11:58:23\"\n",
      "  .. .. .. ..@ assay.used : chr \"RNA\"\n",
      "  .. .. .. ..@ call.string: chr [1:2] \"RunTSNE(object = seu, dims.use = 1:(min(tmp1) - 1), do.fast = TRUE, \" \"    tsne.method = \\\"FIt-SNE\\\", reduction.key = \\\"fitSNE_\\\")\"\n",
      "  .. .. .. ..@ params     :List of 8\n",
      "  .. .. .. .. ..$ reduction     : chr \"pca\"\n",
      "  .. .. .. .. ..$ cells         : chr [1:12396] \"mouseSampleTag5_Aoki_5\" \"mouseSampleTag6_Aoki_21\" \"mouseSampleTag5_Aoki_40\" \"mouseSampleTag6_Aoki_49\" ...\n",
      "  .. .. .. .. ..$ dims          : int [1:5] 1 2 3 4 5\n",
      "  .. .. .. .. ..$ seed.use      : num 1\n",
      "  .. .. .. .. ..$ tsne.method   : chr \"FIt-SNE\"\n",
      "  .. .. .. .. ..$ dim.embed     : num 2\n",
      "  .. .. .. .. ..$ reduction.name: chr \"tsne\"\n",
      "  .. .. .. .. ..$ reduction.key : chr \"fitSNE_\"\n",
      "  .. ..$ RunUMAP.RNA.pca      :Formal class 'SeuratCommand' [package \"SeuratObject\"] with 5 slots\n",
      "  .. .. .. ..@ name       : chr \"RunUMAP.RNA.pca\"\n",
      "  .. .. .. ..@ time.stamp : POSIXct[1:1], format: \"2022-02-10 12:06:57\"\n",
      "  .. .. .. ..@ assay.used : chr \"RNA\"\n",
      "  .. .. .. ..@ call.string: chr \"RunUMAP(seu, dims = dims)\"\n",
      "  .. .. .. ..@ params     :List of 22\n",
      "  .. .. .. .. ..$ dims                : int [1:18] 1 2 3 4 5 6 7 8 9 10 ...\n",
      "  .. .. .. .. ..$ reduction           : chr \"pca\"\n",
      "  .. .. .. .. ..$ assay               : chr \"RNA\"\n",
      "  .. .. .. .. ..$ slot                : chr \"data\"\n",
      "  .. .. .. .. ..$ umap.method         : chr \"uwot\"\n",
      "  .. .. .. .. ..$ return.model        : logi FALSE\n",
      "  .. .. .. .. ..$ n.neighbors         : int 30\n",
      "  .. .. .. .. ..$ n.components        : int 2\n",
      "  .. .. .. .. ..$ metric              : chr \"cosine\"\n",
      "  .. .. .. .. ..$ learning.rate       : num 1\n",
      "  .. .. .. .. ..$ min.dist            : num 0.3\n",
      "  .. .. .. .. ..$ spread              : num 1\n",
      "  .. .. .. .. ..$ set.op.mix.ratio    : num 1\n",
      "  .. .. .. .. ..$ local.connectivity  : int 1\n",
      "  .. .. .. .. ..$ repulsion.strength  : num 1\n",
      "  .. .. .. .. ..$ negative.sample.rate: int 5\n",
      "  .. .. .. .. ..$ uwot.sgd            : logi FALSE\n",
      "  .. .. .. .. ..$ seed.use            : int 42\n",
      "  .. .. .. .. ..$ angular.rp.forest   : logi FALSE\n",
      "  .. .. .. .. ..$ verbose             : logi TRUE\n",
      "  .. .. .. .. ..$ reduction.name      : chr \"umap\"\n",
      "  .. .. .. .. ..$ reduction.key       : chr \"UMAP_\"\n",
      "  ..@ tools       : list()\n"
     ]
    }
   ],
   "source": [
    "###Figure S3F: Extract DEGs between Oligoclonal vs Polyclonal fractions\n",
    "\n",
    "#Specify clone table merged with dLN overlap status\n",
    "file.name  <- \"clone_within_cluster.dLNOL.csv\"\n",
    "dir.input <- \"scTCR_analysis\"\n",
    "dir.name <- \"S2E\"\n",
    "clones_array <- c(\"nonOL\", \"Oligo\", \"Poly\")\n",
    "#Target for DEG analysis\n",
    "Subset1 <- \"Oligo\"\n",
    "Subset2 <- \"Poly\"\n",
    "\n",
    "############################# Processing layer ######################################\n",
    "dir.create(dir.name)\n",
    "\n",
    "#Load clone table merged with dLN overlap status\n",
    "name.input <- str_c(dir.input, file.name, sep = \"/\")\n",
    "d <- read.csv(name.input, header = TRUE)\n",
    "\n",
    "#Call meta.data and cell names\n",
    "meta.data <- as.data.frame(seu@meta.data)\n",
    "tmp_raw <- as.vector(row.names(meta.data))\n",
    "\n",
    "#Integrate OL status into meta.data\n",
    "meta.data$class <- \"nd\"\n",
    "for(i in clones_array){\n",
    "    #Extract clones in each subset\n",
    "    clone_plot <- dplyr::filter(clone_within_cluster, OL == i)\n",
    "    clone_names <- clone_plot$names\n",
    "\n",
    "    #Extract cell names belonging to clones\n",
    "    data_sub <- dplyr::filter(meta.data, clone.id %in% clone_names)\n",
    "    clone.id <- as.vector(row.names(data_sub))\n",
    "    tmp2 = tmp_raw %in% clone.id\n",
    "\n",
    "    #Reflect to meta.data\n",
    "    meta.data$class[tmp2] <- i\n",
    "}\n",
    "\n",
    "#Substitute OL status for \"ident\"\n",
    "ident <- seu@active.ident \n",
    "names.ident <- names(ident)\n",
    "ident <- as.factor(meta.data$class)\n",
    "names(ident) <- names.ident\n",
    "seu@active.ident <- ident\n",
    "\n",
    "#DEG extracton\n",
    "\n",
    "DEG.subset.1 = FindMarkers(seu, ident.1 = Subset1, ident.2 = Subset2,\n",
    "                           test.use=\"wilcox\", only.pos=TRUE, min.pct=0.1,\n",
    "                           features.use = NULL, nthreads = 12,\n",
    "                           adj.p.val.threshold=0.05)\n",
    "name.output <- str_c(Subset1, \"Up.csv\", sep = \".\")\n",
    "name.output <- str_c(dir.name, name.output, sep = \"/\")\n",
    "write.csv(DEG.subset.1, name.output, row.names = TRUE)\n",
    "\n",
    "DEG.subset.2 = FindMarkers(seu, ident.1 = Subset2, ident.2 = Subset1,\n",
    "                           test.use=\"wilcox\", only.pos=TRUE, min.pct=0.1,\n",
    "                           features.use = NULL, nthreads = 12,\n",
    "                           adj.p.val.threshold=0.05)\n",
    "name.output <- str_c(Subset2, \"Up.csv\", sep = \".\")\n",
    "name.output <- str_c(dir.name, name.output, sep = \"/\")\n",
    "write.csv(DEG.subset.2, name.output, row.names = TRUE)\n",
    "\n",
    "#Create DimPlot\n",
    "mBC2 <- subset(seu, idents = c(\"Poly\", \"Oligo\"))\n",
    "genes <- c(\"Tnfrsf4\", \"Fas\", \"Il7r\", \"Bcl6\", \"Xcl1\", \"Ccr7\", \"Tcf7\",\n",
    "           \"Entpd1\", \"Havcr2\", \"Gzmb\")\n",
    "mBC2@meta.data$active.ident <- mBC2@active.ident\n",
    "\n",
    "p <- DotPlot(mBC2, features = genes, cols = c(\"blue\", \"red\"),\n",
    "        dot.min = 0, dot.scale = 4,\n",
    "        scale.by = \"size\", scale.min = NA, scale.max = NA, group.by=meta.data$active.ident) +\n",
    "  theme(axis.title = element_blank(), \n",
    "        axis.text.y = element_text(size=8, colour = 1, family = \"Arial\"),\n",
    "        axis.text.x = element_text(size=8, colour = 1, family = \"Arial\", angle = 90)) +\n",
    "  theme(panel.border = element_rect(fill = NA, size = 0.1),\n",
    "        legend.key.width = unit(0.2, \"cm\"),\n",
    "        legend.key.height = unit(0.2, \"cm\"),\n",
    "        legend.text = element_text(size=6, colour = 1, family = \"Arial\"),\n",
    "        legend.title = element_text(size=6, colour = 1, family = \"Arial\"),\n",
    "        plot.margin = unit(c(0.5, 0, 0, 0), \"cm\"))\n",
    "file.name <- str_c(dir.name, \"DEG.OligovsPoly.tiff\", sep=\"/\")\n",
    "save_plot(file = file.name, plot(p), device=\"tiff\", \n",
    "          units=\"in\", dpi = 600, base_width = 3.5, base_height = 1.6, limitsize=FALSE)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
