{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "25a3eb7b",
   "metadata": {},
   "source": [
    "### Load libraries and functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a9a52127",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load libraries\n",
    "library(slingshot)\n",
    "library(ggsci)\n",
    "library(scales)\n",
    "library(tidyr)\n",
    "library(Seurat)\n",
    "library(Rcpp)\n",
    "library(parallelDist)\n",
    "library(stringr)\n",
    "library(viridis)\n",
    "library(ggplot2)\n",
    "library(cowplot)\n",
    "library(cluster)\n",
    "library(data.table)\n",
    "library(foreach)\n",
    "library(doParallel)\n",
    "library(proxy)\n",
    "library(ComplexHeatmap)\n",
    "library(circlize)\n",
    "library(igraph)\n",
    "library(qvalue)\n",
    "library(dplyr)\n",
    "library(viridis)\n",
    "library(VGAM)\n",
    "library(forcats)\n",
    "library(grDevices)\n",
    "library(graphics)\n",
    "library(RColorBrewer)\n",
    "library(pheatmap)\n",
    "library(Cairo)\n",
    "library(reshape2)\n",
    "library(R.utils)\n",
    "set.seed(seed = 42)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "54558b75",
   "metadata": {},
   "outputs": [],
   "source": [
    "#load basic functions\n",
    "createEmptyDf = function( nrow, ncol, colnames = c() ){\n",
    "  data.frame( matrix( vector(), nrow, ncol, dimnames = list( c(), colnames ) ) )\n",
    "}\n",
    "\n",
    "tableread_fast = function(i, header=TRUE, quote=\"\", sep=\",\"){\n",
    "  tmp = fread(i, header=header, sep=sep, quote=quote, nThread=32)\n",
    "  tmp = as.data.frame(tmp)\n",
    "  return(tmp)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a7221f3f",
   "metadata": {},
   "source": [
    "### Create Seurat object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "73dd4b1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#load data: aCD4 SCT\n",
    "#Put gene list in the target panel in the working directory (\"BD_genes_plus10\")\n",
    "\n",
    "###Input layer\n",
    "set.seed(seed = 42)\n",
    "input_dir <- \"data/SCT_rawdata\"\n",
    "sample.name <- \"aCD4\"\n",
    "\n",
    "################################ Processing layer ##########################################\n",
    "#Create input file name\n",
    "input_name <- str_c(input_dir, \"matrix_inflection_demulti_\", sep = \"/\") %>% str_c(sample.name, \".txt\", sep = \"\") #Name of dataset1\n",
    "\n",
    "#Unzip input file\n",
    "compressed.name <- str_c(input_name, \".gz\", sep = \"\")\n",
    "gunzip(compressed.name)\n",
    "\n",
    "#Create Seurat object\n",
    "matrix <- tableread_fast(input_name, header = TRUE, quote=\"\", sep=\"\\t\")\n",
    "row.names(matrix) <- matrix$V1\n",
    "matrix <- dplyr::select(matrix, -V1)\n",
    "seu1 <- CreateSeuratObject(counts=matrix, project = \"seu\", min.cells = 3, min.features = 10)\n",
    "gzip(str_remove(compressed.name, \".gz\"))\n",
    "unlink(str_remove(compressed.name, \".gz\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "10583b5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#load data: Control/aPDL1 SCT\n",
    "#Put gene list in the target panel in the working directory (\"BD_genes_plus10\")\n",
    "\n",
    "###Input layer\n",
    "set.seed(seed = 42)\n",
    "\n",
    "sample.name <- \"Cont_aPDL1\"\n",
    "\n",
    "################################ Processing layer ##########################################\n",
    "#Create input file name\n",
    "input_name <- str_c(input_dir, \"matrix_inflection_demulti_\", sep = \"/\") %>% str_c(sample.name, \".txt\", sep = \"\") #Name of dataset1\n",
    "\n",
    "#Unzip input file\n",
    "compressed.name <- str_c(input_name, \".gz\", sep = \"\")\n",
    "gunzip(compressed.name)\n",
    "\n",
    "#Create Seurat object\n",
    "matrix <- tableread_fast(input_name, header = TRUE, quote=\"\", sep=\"\\t\")\n",
    "row.names(matrix) <- matrix$V1\n",
    "matrix <- dplyr::select(matrix, -V1)\n",
    "seu2 <- CreateSeuratObject(counts=matrix, project = \"seu\", min.cells = 3, min.features = 10)\n",
    "gzip(str_remove(compressed.name, \".gz\"))\n",
    "unlink(str_remove(compressed.name, \".gz\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed3cd178",
   "metadata": {},
   "source": [
    "### Preprocessing for excluding check non-T cell contamination"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "be77506b",
   "metadata": {},
   "outputs": [],
   "source": [
    "###Function: perform Seurat pipeline: Normalizing, Scaling, PCA, JackStraw, dimentional reduction, find clusters\n",
    "##Input: seurat object\n",
    "##Output: seurat object, JackStraw Plot\n",
    "\n",
    "Preprocess <- function(seu, dir.name, sample.name){\n",
    "    # normalizing data\n",
    "    seu = NormalizeData(object = seu, scale.factor=1000000)\n",
    "    VariableFeatures(seu) <- rownames(seu[[\"RNA\"]]) #Variable features: all genes\n",
    "    ngenes <- length(x = seu@assays$RNA@var.features)\n",
    "    all.genes <- rownames(seu)\n",
    "    seu = ScaleData(object = seu, vars.to.regress = c(\"nCount_RNA\"), features = all.genes) \n",
    "\n",
    "    # perform PCA\n",
    "    seu = RunPCA(object = seu, features = seu@assays$RNA@var.features, npcs = 50)\n",
    "\n",
    "    #JackStraw\n",
    "    seu = JackStraw(object = seu, num.replicate = 100, dims = 50)\n",
    "    seu <- ScoreJackStraw(object = seu, dims = 1:50, score.thresh = 0.05)\n",
    "    file.name <- str_c(dir.name, sample.name, sep='/') %>% str_c(\"_JackStraw.png\", sep='')\n",
    "    png(file.name, width = 1250, height = 500)\n",
    "    p <- JackStrawPlot(object = seu, dims = 1:50)\n",
    "    plot(p)\n",
    "    dev.off()\n",
    "\n",
    "    #Determine PCs used for clustering/tSNE analysis (dims.use)\n",
    "    #Extract PCs which fulfill the pvalue threshold\n",
    "    tmp = as.data.frame(seu@reductions$pca@jackstraw@overall.p.values)\n",
    "    tmp1 = tmp[tmp$Score>0.05,1]\n",
    "    dims= c(1:(min(tmp1)-1))\n",
    "\n",
    "    ##Perform dimentional reduction\n",
    "    seu <- RunUMAP(object = seu, dims = 1:(min(tmp1)-1))\n",
    "\n",
    "    ##Find clusters\n",
    "    seu = FindNeighbors(seu, reduction = \"pca\", dims = dims, force.recalc = TRUE)\n",
    "    \n",
    "    return(seu)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "0d75c3f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "###Function: DimPlot with sample origin\n",
    "##Input: seurat object\n",
    "##Output: DimplotStraw Plot\n",
    "\n",
    "DimOrigin <- function(seu, dir.name, sample.name, red.use, resol){\n",
    "    p1 = DimPlot(object = seu, reduction = red.use, label = TRUE, label.size = 10, pt.size = 0.5) +\n",
    "      theme(axis.title.x = element_text(size=10, family = \"Arial\"), \n",
    "            axis.title.y = element_text(size=10, family = \"Arial\"), \n",
    "            axis.text.x = element_text(size=10, colour = 1, family = \"Arial\"), \n",
    "            axis.text.y = element_text(size = 10, colour = 1, family = \"Arial\")) +\n",
    "      theme(panel.border = element_rect(fill = NA, size = 1)) \n",
    "    p2 = DimPlot(object = seu, reduction = red.use, label = FALSE, label.size = 10, pt.size = 0.5, group.by = \"orig.ident\") +\n",
    "      theme(axis.title.x = element_text(size=10, family = \"Arial\"),\n",
    "            axis.title.y = element_text(size=10, family = \"Arial\"),\n",
    "            axis.text.x = element_text(size=10, colour = 1, family = \"Arial\"),\n",
    "            axis.text.y = element_text(size = 10, colour = 1, family = \"Arial\")) +\n",
    "      theme(panel.border = element_rect(fill = NA, size = 1)) \n",
    "\n",
    "    legend1 <- cowplot::get_legend(p1)\n",
    "    legend2 <- cowplot::get_legend(p2)\n",
    "    p1 = p1 + theme(legend.position = 'none')\n",
    "    p2 = p2 + theme(legend.position = 'none')\n",
    "    file.name <- str_c(dir.name, sample.name, sep='/') %>% str_c(red.use, \"reso\", resol, \"png\", sep='.')\n",
    "    save_plot(file = file.name, plot_grid(p1, legend1, p2, legend2, ncol=2, nrow=2), device=\"png\", \n",
    "              units=\"in\", dpi = 600, base_width = 10, base_height = 10, limitsize=FALSE)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "827335b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "###Function: Find marker genes with marker heatmap\n",
    "##Input: seurat object\n",
    "##Output: Marker gene table and marker gene heatmap plot\n",
    "\n",
    "MarkerHeatmap <- function(seu, dir.name, sample.name, resol){\n",
    "    seu.markers = FindAllMarkers(seu, verbose = TRUE, test.use=\"wilcox\", only.pos=TRUE, min.pct=0.1, features.use = NULL, return.thresh=0.05)\n",
    "\n",
    "    #Create heatmap with top10 marker genes\n",
    "    top10 = seu.markers %>% group_by(cluster) %>% top_n(10, avg_log2FC)\n",
    "    top10 = as.data.frame(top10)\n",
    "    top10  = top10 [!duplicated(top10$gene),]\n",
    "    top10 = top10 %>% arrange(desc(avg_log2FC))  %>% arrange(cluster)\n",
    "    top10 = as.data.frame(top10)\n",
    "    file.name <- str_c(dir.name, sample.name, sep='/') %>% str_c(\"marker_res\", resol, \"png\", sep='.')\n",
    "    p <- DoHeatmap(seu, features = top10$gene, disp.min = -2.5, disp.max = 2.5, size = 8)\n",
    "    ggsave(file = file.name, plot = p, device=\"png\", units=\"in\", dpi = 300,\n",
    "           width = 20, height = 20, limitsize=FALSE)\n",
    "\n",
    "    seu.markers$cluster = as.numeric(seu.markers$cluster)\n",
    "    seu.markers = seu.markers %>% arrange(desc(avg_log2FC))  %>% arrange(cluster)\n",
    "    seu.markers = as.data.frame(seu.markers)\n",
    "    file.name <- str_c(dir.name, sample.name, sep='/') %>% str_c(\"ALLmarkers_minpct0.1_Adj_p0.05.txt\", sep='')\n",
    "    fwrite(seu.markers, file.name, row.names=F, col.names=T, sep=\"\\t\", quote=F)\n",
    "    file.name <- str_c(dir.name, sample.name, sep='/') %>% str_c(\"ALLmarkers_minpct0.1_Adj_p0.05.rda\", sep='')\n",
    "    save(seu.markers, file=file.name)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "20ac7fdf",
   "metadata": {},
   "outputs": [],
   "source": [
    "###Function: Create Scatter and Violin plot for selected genes\n",
    "##Input: seurat object, gene list\n",
    "##Output: Scatter plot, Violin plot\n",
    "\n",
    "ScatterViolin <- function(seu, dir.name, sample.name, red.use, tmps, tmp_names){\n",
    "    for (i in 1:length(tmp_names)){\n",
    "        #Call gene list\n",
    "        tmp <- tmps[[i]]\n",
    "        tmp_name <- tmp_names[i]\n",
    "  \n",
    "        #Scatter Plot\n",
    "        file.name <- str_c(dir.name, sample.name, sep='/') %>% str_c(\"Scatter\", tmp_name, \"png\", sep='.')\n",
    "        png(file.name, width = 1536, height = 240)\n",
    "        p <- FeaturePlot(seu, features = tmp, ncol = 6, order = TRUE,\n",
    "                        reduction = red.use, dims=c(1,2), cols = c(\"grey\", \"red\"), pt.size = 0.2)\n",
    "        plot(p)\n",
    "        dev.off()\n",
    "        \n",
    "        #Violin Plot\n",
    "        file.name <- str_c(dir.name, sample.name, sep='/') %>% str_c(\"Violin\", tmp_name, \"png\", sep='.')\n",
    "        png(file.name, width = 1536, height = 240)\n",
    "        p <- VlnPlot(seu, features = tmp, ncol = 6, pt.size = 0.1)\n",
    "        plot(p)\n",
    "        dev.off()\n",
    "    }\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81f2cb84",
   "metadata": {},
   "source": [
    "### Quality Check"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "15b70688",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Merge two datasets\n",
    "seu <- merge(x=seu1, y = seu2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "4bce64dd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Scale for 'x' is already present. Adding another scale for 'x', which will\n",
      "replace the existing scale.\n",
      "\n",
      "Picking joint bandwidth of 0.0332\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Picking joint bandwidth of 2.91\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "###Gene filtering\n",
    "\n",
    "##Input layer\n",
    "sample.name <- \"Pooled\"\n",
    "dir.name=\"results/preprocess/Seurat_plot\"\n",
    "gene.list.name <- \"BD_genes_plus10.txt\"\n",
    "\n",
    "########################## Processing layer ###############################\n",
    "dir.name <- str_c(dir.name, sep = \"/\") %>% str_c(\"Stat\", sep = \"/\")\n",
    "dir.create(dir.name, recursive = TRUE)\n",
    "\n",
    "##load gene list of BD target panel\n",
    "BD_genes <- read.table(gene.list.name, header = TRUE)\n",
    "BD_genes <- as.vector(BD_genes$Genesymbol)\n",
    "\n",
    "##Chose BD target gene list for downstream analysis\n",
    "res1 = seu@assays$RNA@counts\n",
    "res1 =res1[res1@Dimnames[[1]] %in% BD_genes,]\n",
    "ngenes <- length(res1@Dimnames[[1]])\n",
    "seu@assays$RNA@counts <- res1\n",
    "seu@assays$RNA@data <- res1\n",
    "    \n",
    "#Remove doublet and Pmel cells\n",
    "seu <- subset(seu, subset = orig.ident %in% c(\"mouseSampleTag1\", \"mouseSampleTag2\",\n",
    "                                                \"mouseSampleTag5\", \"mouseSampleTag6\"))\n",
    "\n",
    "#Scatter plot for gene/read count\n",
    "file.name=str_c(dir.name, sample.name, sep = \"/\") %>% str_c(\"Reads.Genes.png\", sep='.')\n",
    "png(file.name, width = 512, height = 400)\n",
    "p <- FeatureScatter(object = seu, feature1 = \"nCount_RNA\", feature2 = \"nFeature_RNA\", pt.size = 0.3) +\n",
    "      geom_hline(yintercept=10) +\n",
    "      scale_x_log10()\n",
    "plot(p)\n",
    "dev.off()\n",
    "\n",
    "#RidgePlot for gene/read count, \n",
    "file.name=str_c(dir.name, sample.name, sep = \"/\") %>% str_c(\"Reads.png\", sep='.')\n",
    "png(file.name, width = 512, height = 400)\n",
    "p <- RidgePlot(object = seu, features = \"nCount_RNA\", group.by=\"orig.ident\", ncol = 1) +\n",
    "      scale_x_log10()\n",
    "plot(p)\n",
    "dev.off()\n",
    "\n",
    "file.name=str_c(dir.name, sample.name, sep = \"/\") %>% str_c(\"Genes.png\", sep='.')\n",
    "png(file.name, width = 512, height = 400)\n",
    "p <- RidgePlot(object = seu, features = \"nFeature_RNA\", group.by=\"orig.ident\", ncol = 1)\n",
    "plot(p)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b02c8fe",
   "metadata": {},
   "source": [
    "###  Initial analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "58d52510",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Regressing out nCount_RNA\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  Pdcd1, Lag3, Cxcr6, Lgals1, Ctla4, Gzmb, Bcl2a1a, Nkg7, Klrc1, S100a10 \n",
      "\t   Havcr2, Ifng, Ctsd, Lgals3, Cd8a, Icos, Cd48, Ccl5, Ccr2, Cst7 \n",
      "\t   Gzmk, Ybx3, Itgb2, Ccr5, Cd3e, Klrk1, Tigit, Tnfrsf9, Cd5, Cd3g \n",
      "Negative:  Tcf7, Il7r, Ccr7, Sell, Btg1, Lef1, Il4ra, Il6ra, H2-K1, Txk \n",
      "\t   Arl4c, Cd14, Adgre1, Fcgr3, Cd300a, Pik3ip1, Cxcl2, Cd63, Trem2, Il1rn \n",
      "\t   Nt5e, Tnfsf13, Itk, Ly86, Cd36, Slc11a1, Clec4d, Mgst1, Ccl2, Cxcl16 \n",
      "PC_ 2 \n",
      "Positive:  Lyz2, Cd14, H2-Aa, Fcgr3, Ccl6, H2-Eb1, Ly86, Il1rn, Ifi204, Cxcl2 \n",
      "\t   Adgre1, H2-Ab1, C1qb, Slc11a1, Itgam, Ccr1, Ccl2, Cd63, Cxcl16, Trem2 \n",
      "\t   Ifitm3, Lgals3, Cd300a, Ccl9, C1qa, Nlrp3, Clec4d, Tnfsf13, Cd36, Thbs1 \n",
      "Negative:  Ccr7, Tcf7, Itk, H2-K1, Il7r, Stat3, Xcl1, Stat4, Trbc2, Sell \n",
      "\t   Il4ra, Nfkb1, Ptprc, Tnfrsf4, Stat5a, Lef1, Txk, Tnfsf8, Cd3d, Kit \n",
      "\t   Cxcr5, Nt5e, Cblb, Tfrc, Ighm, Btg1, Nfkb2, Mcm4, Tgfb1, Tnfrsf18 \n",
      "PC_ 3 \n",
      "Positive:  Tnfrsf4, Cd74, Xcl1, Tnfrsf9, Irf8, Nrp1, Il1r2, Lag3, Ccr7, Cd200 \n",
      "\t   Tnfrsf18, Kit, Cd9, Il2ra, Ccr6, Junb, H2-Ab1, Dusp1, H2-Eb1, Tnfsf8 \n",
      "\t   Cd160, Stat3, Ccr8, Csf2, Pdcd1, Anxa5, H2-Aa, Bcl2a1a, Slc11a1, Nfkb2 \n",
      "Negative:  Ccl5, Gzmk, Itgax, Ifngr1, Gzma, Cxcr3, Gzmb, Klrg1, Dusp2, Ccr2 \n",
      "\t   Gimap7, Gzmm, Fasl, Btg1, Lat, Arl4c, Cxcr4, Igkc, Itga4, Eomes \n",
      "\t   Nkg7, Rgs1, Nod1, Ifit3b, Ifit3, Selplg, Entpd1, Cd72, Fyn, Lgals3 \n",
      "PC_ 4 \n",
      "Positive:  Mki67, Tyms, Top2a, Ube2c, Junb, Mcm4, Havcr2, Mcm2, Ccl3, Pask \n",
      "\t   Ccl4, Itgae, Entpd1, Gzmb, Trib2, Dusp1, Sell, Tmem97, Il12rb1, Il6ra \n",
      "\t   Ccr5, Cd247, Pcna, Lat, Ifng, Aurkb, Myc, Pdia4, Stat1, Tfrc \n",
      "Negative:  Klrc1, Il18rap, Ccl5, Cd44, Cd6, Xcl1, Cxcr3, Itgax, Klra7, Rora \n",
      "\t   Nfkb2, Tnfrsf25, Tnfrsf1b, Klrk1, Klra6, Igkc, Tcf7, Eomes, Ctsw, Gzmm \n",
      "\t   Stat4, Cxcr4, Tigit, Fyn, Cd52, Kit, Il12rb2, Nfkb1, Nkg7, Cd9 \n",
      "PC_ 5 \n",
      "Positive:  Tnfrsf9, Tnfrsf4, Il1r2, Irf8, Il18rap, Klrc1, Spp1, Kit, Cd200, Il2ra \n",
      "\t   Cx3cr1, Csf2, Prf1, Nrp1, Ccr8, Ccl3, Tgfb1, Il1rl1, Igkc, Itgax \n",
      "\t   Rora, Csf1, Itga4, Bcl2a1a, Lgals3, Lag3, Ccr6, Cd9, Havcr2, Lat2 \n",
      "Negative:  Ifit3b, Ifit3, Isg15, Rsad2, Ifit1, Irf7, Lgals9, Mx1, Cmpk2, Ddx58 \n",
      "\t   Cd7, Usp18, Sell, Cxcl10, Klra7, Stat1, Selplg, Klra6, Ighm, Ifi211 \n",
      "\t   Junb, Bin2, Fth1, Cd69, Cxcr4, Dusp2, Ifi204, Dusp1, Ifngr1, Cxcr3 \n",
      "\n",
      "Warning message:\n",
      "“Removed 15915 rows containing missing values (geom_point).”\n",
      "Warning message:\n",
      "“The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric\n",
      "To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'\n",
      "This message will be shown once per session”\n",
      "11:42:13 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "11:42:13 Read 35544 rows and found 25 numeric columns\n",
      "\n",
      "11:42:13 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "11:42:13 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "11:42:17 Writing NN index file to temp file /tmp/Rtmp0EaSh0/fileb1671c6f\n",
      "\n",
      "11:42:17 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "11:42:30 Annoy recall = 100%\n",
      "\n",
      "11:42:30 Commencing smooth kNN distance calibration using 1 thread\n",
      "\n",
      "11:42:31 Initializing from normalized Laplacian + noise\n",
      "\n",
      "11:42:32 Commencing optimization for 200 epochs, with 1583714 positive edges\n",
      "\n",
      "11:42:46 Optimization finished\n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 35544\n",
      "Number of edges: 956469\n",
      "\n",
      "Running Louvain algorithm...\n",
      "Maximum modularity in 10 random starts: 0.8151\n",
      "Number of communities: 17\n",
      "Elapsed time: 12 seconds\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "1 singletons identified. 16 final clusters.\n",
      "\n",
      "Warning message in grid.Call(C_stringMetric, as.graphicsAnnot(x$label)):\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_stringMetric, as.graphicsAnnot(x$label)):\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_stringMetric, as.graphicsAnnot(x$label)):\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_stringMetric, as.graphicsAnnot(x$label)):\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_stringMetric, as.graphicsAnnot(x$label)):\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_stringMetric, as.graphicsAnnot(x$label)):\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_stringMetric, as.graphicsAnnot(x$label)):\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_stringMetric, as.graphicsAnnot(x$label)):\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_stringMetric, as.graphicsAnnot(x$label)):\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_stringMetric, as.graphicsAnnot(x$label)):\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_stringMetric, as.graphicsAnnot(x$label)):\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_stringMetric, as.graphicsAnnot(x$label)):\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_stringMetric, as.graphicsAnnot(x$label)):\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_stringMetric, as.graphicsAnnot(x$label)):\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Calculating cluster 0\n",
      "\n",
      "Calculating cluster 1\n",
      "\n",
      "Calculating cluster 2\n",
      "\n",
      "Calculating cluster 3\n",
      "\n",
      "Calculating cluster 4\n",
      "\n",
      "Calculating cluster 5\n",
      "\n",
      "Calculating cluster 6\n",
      "\n",
      "Calculating cluster 7\n",
      "\n",
      "Calculating cluster 8\n",
      "\n",
      "Calculating cluster 9\n",
      "\n",
      "Calculating cluster 10\n",
      "\n",
      "Calculating cluster 11\n",
      "\n",
      "Calculating cluster 12\n",
      "\n",
      "Calculating cluster 13\n",
      "\n",
      "Calculating cluster 14\n",
      "\n",
      "Calculating cluster 15\n",
      "\n"
     ]
    }
   ],
   "source": [
    "###Run Seurat pipeline (dimentional reduction ~ determining resolution) with iterative process\n",
    "\n",
    "###Input layer \n",
    "sample.name <- \"Pooled\"\n",
    "dir.name <- \"results/preprocess/Seurat_plot\"\n",
    "dir.seurat <- \"results/preprocess\"\n",
    "red.use <- \"umap\"\n",
    "resol <- 1.0\n",
    "\n",
    "#For lineage check\n",
    "tmp_1 = c(\"Trbc2\", \"Cd3e\", \"Cd8a\", \"Cd4\", \"Cd14\", \"Lyz2\")\n",
    "tmps <- list(tmp_1)\n",
    "tmp_names <- c(\"lineage\")\n",
    "\n",
    "########################## Processing layer #############################\n",
    "### Start processing ###\n",
    "seu <- Preprocess(seu, dir.name, sample.name)\n",
    "    \n",
    "#Find clusters\n",
    "seu <- FindClusters(object = seu, resolution =resol)\n",
    "    \n",
    "#DimPlot with sample origin\n",
    "DimOrigin(seu, dir.name, sample.name, red.use, resol)\n",
    "    \n",
    "#Marker gene extraction and create marker gene heatmap\n",
    "MarkerHeatmap(seu, dir.name, sample.name, resol)\n",
    "    \n",
    "#Check contamination\n",
    "ScatterViolin(seu, dir.name, sample.name, red.use, tmps, tmp_names)\n",
    "        \n",
    "#Output Seurat object in 1st analysis\n",
    "file.name <- str_c(dir.seurat, sample.name, sep = \"/\") %>% str_c(\".rda\", sep='')\n",
    "save(seu, file=file.name)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "753f88a4",
   "metadata": {},
   "source": [
    "### 2nd analysis with removing contaminant non-T cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "f6026abc",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Remove contaminant clusters\n",
    "seu <- subset(seu, idents = c(0, 1, 2, 3, 4, 5, 6, 8, 9, 10, 11, 13)) #7: NK cells, 12,15:Macrophage, 14:Treg?\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "d19202d8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Regressing out nCount_RNA\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "Warning message in PrepDR(object = object, features = features, verbose = verbose):\n",
      "“The following 1 features requested have zero variance (running reduction without them): Arg1”\n",
      "PC_ 1 \n",
      "Positive:  Ccr7, Tcf7, Il7r, Sell, Btg1, Lef1, Il4ra, H2-K1, Itk, Txk \n",
      "\t   Il6ra, Arl4c, Nt5e, Pik3ip1, Stat3, Trib2, Bcl2, Cxcr5, Bach2, Itgae \n",
      "\t   Stat5a, Myc, Fas, Birc3, Ikbkb, Tnfsf8, H2-Ob, Socs1, Ighm, Stat4 \n",
      "Negative:  Pdcd1, Gzmb, Cxcr6, Lgals1, Lag3, Ctla4, Lgals3, Nkg7, Ccl5, Bcl2a1a \n",
      "\t   Ccr2, Ctsd, Klrc1, S100a10, Havcr2, Ifng, Gzmk, Ccr5, Icos, Cd8a \n",
      "\t   Cd48, Itgb2, Fasl, Klrk1, Cst7, Gimap7, Tigit, Ccl4, Ybx3, Casp1 \n",
      "PC_ 2 \n",
      "Positive:  Ccl5, Gzmk, Itgax, Gzma, Gzmb, Ifngr1, Klrg1, Ccr2, Cxcr3, Gimap7 \n",
      "\t   Arl4c, Ifit3b, Btg1, Gzmm, Fasl, Ifit3, Dusp2, Cxcr4, Igkc, Entpd1 \n",
      "\t   Cmklr1, Ccr5, Lgals3, Itga4, Irf7, Lat, Nod1, Rgs1, Cd72, Eomes \n",
      "Negative:  Tnfrsf4, Xcl1, Tnfrsf9, Nrp1, Lag3, Irf8, Cd74, Tnfrsf18, Kit, Ccr7 \n",
      "\t   Il1r2, Ccr6, Pdcd1, Cd200, Cd9, Cd160, Bcl2a1a, Il2ra, Tnfsf8, Ccr8 \n",
      "\t   Ctla4, Stat3, Junb, Csf2, Cst7, Ikzf2, Nfkb2, Ybx3, Dusp1, Nfkb1 \n",
      "PC_ 3 \n",
      "Positive:  Junb, Mki67, Tyms, Havcr2, Ube2c, Ccl3, Top2a, Ccl4, Mcm4, Dusp1 \n",
      "\t   Mcm2, Sell, Entpd1, Ccr5, Ifng, Pask, Gzmb, Il12rb1, Lat2, Cd247 \n",
      "\t   Cd7, Stat1, Tmem97, Itgae, Il6ra, Trib2, Myc, Irf8, Pdia4, Pcna \n",
      "Negative:  Cd6, Klrc1, Il18rap, Cxcr3, Itgax, Igkc, Cd44, Tnfrsf25, Ccl5, Rora \n",
      "\t   Nfkb2, H2-K1, Tcf7, Gzmm, Fyn, Tnfrsf1b, Eomes, Xcl1, Stat4, Icos \n",
      "\t   Tigit, Cd28, Kit, Bcl2a1a, Cd52, Klrk1, Nkg7, Nfkb1, Ccr6, Cd9 \n",
      "PC_ 4 \n",
      "Positive:  Ifit3b, Ifit3, Ifit1, Isg15, Rsad2, Irf7, Mx1, Cmpk2, Lgals9, Ddx58 \n",
      "\t   Usp18, Cxcl10, Cd7, Selplg, Ifi211, Lck, Junb, Cxcr3, Stat1, Sell \n",
      "\t   Cd8b1, Cd3g, Cd2, Btg1, Tcf7, Ptprc, Fth1, Ighm, Bin2, Lat \n",
      "Negative:  Tnfrsf9, Irf8, Ccl3, Il1r2, Spp1, Il2ra, Il18rap, Prf1, Klrc1, Lat2 \n",
      "\t   Ccl9, Fcer1g, Cx3cr1, Lgals3, Tnfrsf4, Havcr2, Klrg1, Itgax, Cd200, Csf2 \n",
      "\t   Gzmb, Kit, Itga4, Entpd1, Il12rb2, Csf1, Lyn, Ccl5, Tgfb1, Igkc \n",
      "PC_ 5 \n",
      "Positive:  Mki67, Top2a, Tyms, Mcm2, Ube2c, Mcm4, Pcna, Eomes, Pask, Tmem97 \n",
      "\t   Gzmk, Aurkb, Rpn2, Gapdh, Cxcr3, Hprt, Cxcr4, Lat, Mapk1, Tfrc \n",
      "\t   Klrg1, Igkc, Bin2, Lamp1, Pdia4, Fyn, Gimap7, Cd28, Ighm, Cd5 \n",
      "Negative:  Ifit3b, Ifit3, Ifit1, Rsad2, Mx1, Isg15, Cmpk2, Usp18, Csf1, Ifitm3 \n",
      "\t   Irf7, Ifi204, Lat2, Ifitm2, Ddx58, Ccl9, Lgals3, Prf1, Fcer1g, Cd86 \n",
      "\t   Il1r2, Il12rb2, Tnfrsf9, Btg1, Irf8, Lag3, Klrk1, Il2ra, Lilrb4a, Nrp1 \n",
      "\n",
      "Warning message:\n",
      "“Removed 16078 rows containing missing values (geom_point).”\n",
      "12:04:37 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "12:04:37 Read 32870 rows and found 26 numeric columns\n",
      "\n",
      "12:04:37 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "12:04:37 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "12:04:40 Writing NN index file to temp file /tmp/Rtmp0EaSh0/fileb5849bff4\n",
      "\n",
      "12:04:40 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "12:04:52 Annoy recall = 100%\n",
      "\n",
      "12:04:52 Commencing smooth kNN distance calibration using 1 thread\n",
      "\n",
      "12:04:53 Initializing from normalized Laplacian + noise\n",
      "\n",
      "12:04:54 Commencing optimization for 200 epochs, with 1459918 positive edges\n",
      "\n",
      "12:05:07 Optimization finished\n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 32870\n",
      "Number of edges: 811262\n",
      "\n",
      "Running Louvain algorithm...\n",
      "Maximum modularity in 10 random starts: 0.7951\n",
      "Number of communities: 15\n",
      "Elapsed time: 10 seconds\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "1 singletons identified. 14 final clusters.\n",
      "\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Calculating cluster 0\n",
      "\n",
      "Calculating cluster 1\n",
      "\n",
      "Calculating cluster 2\n",
      "\n",
      "Calculating cluster 3\n",
      "\n",
      "Calculating cluster 4\n",
      "\n",
      "Calculating cluster 5\n",
      "\n",
      "Calculating cluster 6\n",
      "\n",
      "Calculating cluster 7\n",
      "\n",
      "Calculating cluster 8\n",
      "\n",
      "Calculating cluster 9\n",
      "\n",
      "Calculating cluster 10\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Calculating cluster 11\n",
      "\n",
      "Calculating cluster 12\n",
      "\n",
      "Calculating cluster 13\n",
      "\n"
     ]
    }
   ],
   "source": [
    "###Run Seurat pipeline (dimentional reduction ~ determining resolution) with iterative process\n",
    "\n",
    "###Input layer \n",
    "sample.name <- \"Pooled\"\n",
    "dir.name <- \"results/preprocess/Seurat_plot_2nd\"\n",
    "red.use <- \"umap\"\n",
    "resol <- 1.0\n",
    "\n",
    "#For lineage check\n",
    "tmp_1 = c(\"Trbc2\", \"Cd3e\", \"Cd8a\", \"Cd4\", \"Cd14\", \"Lyz2\")\n",
    "tmps <- list(tmp_1)\n",
    "tmp_names <- c(\"lineage\")\n",
    "\n",
    "########################## Processing layer #############################\n",
    "### Start processing ###\n",
    "dir.create(dir.name)\n",
    "seu <- Preprocess(seu, dir.name, sample.name)\n",
    "    \n",
    "#Find clusters\n",
    "seu <- FindClusters(object = seu, resolution =resol)\n",
    "    \n",
    "#DimPlot with sample origin\n",
    "DimOrigin(seu, dir.name, sample.name, red.use, resol)\n",
    "    \n",
    "#Marker gene extraction and create marker gene heatmap\n",
    "MarkerHeatmap(seu, dir.name, sample.name, resol)\n",
    "    \n",
    "#Check contamination\n",
    "ScatterViolin(seu, dir.name, sample.name, red.use, tmps, tmp_names)\n",
    "        \n",
    "#Output Seurat object in 2nd analysis\n",
    "file.name <- str_c(dir.seurat, sample.name, sep = \"/\") %>% str_c(\"_2nd.rda\", sep='')\n",
    "save(seu, file=file.name)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57f1b642",
   "metadata": {},
   "source": [
    "### Determine clustering resolution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "752f65f7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 32870\n",
      "Number of edges: 811262\n",
      "\n",
      "Running Louvain algorithm...\n",
      "Maximum modularity in 10 random starts: 0.8628\n",
      "Number of communities: 10\n",
      "Elapsed time: 10 seconds\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "1 singletons identified. 9 final clusters.\n",
      "\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Calculating cluster 0\n",
      "\n",
      "Calculating cluster 1\n",
      "\n",
      "Calculating cluster 2\n",
      "\n",
      "Calculating cluster 3\n",
      "\n",
      "Calculating cluster 4\n",
      "\n",
      "Calculating cluster 5\n",
      "\n",
      "Calculating cluster 6\n",
      "\n",
      "Calculating cluster 7\n",
      "\n",
      "Calculating cluster 8\n",
      "\n"
     ]
    }
   ],
   "source": [
    "###Visualize clusters\n",
    "\n",
    "###Input layer \n",
    "sample.name <- \"Pooled\"\n",
    "dir.name <- \"results/preprocess/Seurat_plot_Clustering\"\n",
    "red.use <- \"umap\"\n",
    "resol <- 0.5\n",
    "\n",
    "########################## Processing layer #############################\n",
    "dir.create(dir.name)\n",
    "    \n",
    "#Find clusters\n",
    "seu <- FindClusters(object = seu, resolution =resol)\n",
    "    \n",
    "#DimPlot with sample origin\n",
    "DimOrigin(seu, dir.name, sample.name, red.use, resol)\n",
    "    \n",
    "#Marker gene extraction and create marker gene heatmap\n",
    "MarkerHeatmap(seu, dir.name, sample.name, resol)\n",
    "    \n",
    "#Output Seurat object in 2nd analysis\n",
    "file.name <- str_c(dir.seurat, sample.name, sep = \"/\") %>% str_c(\"_clust.rda\", sep='')\n",
    "save(seu, file=file.name)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b885fbf",
   "metadata": {},
   "source": [
    "### Pseudotime Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "dca1c3c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "load(\"results/preprocess/Pooled_clust.rda\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "1e19eeb0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Pseudotime analysis\n",
    "#seu.sub: create pseudotime for subset of clusters\n",
    "#red.use: dimentional reduction for plotting pseudotime\n",
    "Pseudotime.Min <- function(seu, seu.sub, name, dims, start, dir.name){\n",
    "  #Create lineages and pseudotime\n",
    "  lineages = slingshot(seu.sub@reductions$pca@cell.embeddings[,dims], \n",
    "                       clusterLabels = seu.sub@active.ident, maxit=1000, start.clus=start)\n",
    "  pseudotime=slingPseudotime(lineages, na = TRUE)\n",
    "  \n",
    "  #calculate pseudotime for each lineage\n",
    "  for (i in c(1:length(lineages@metadata$lineages))){\n",
    "    pseudotime[,i]=pseudotime[,i]/max(pseudotime[,i][!is.na(pseudotime[,i])])\n",
    "  }\n",
    "  if (length(lineages@metadata$lineages) == 1) {\n",
    "    pseudotime1 = pseudotime[,1]\n",
    "  } else {\n",
    "    pseudotime1 = Matrix::rowMeans(pseudotime, na.rm=TRUE)\n",
    "  }\n",
    "    \n",
    "    print(SlingshotDataSet(lineages))\n",
    "  \n",
    "  #Add to MetaData\n",
    "  seu2 = AddMetaData(object = seu, metadata = pseudotime1, col.name = \"pseudotime\")\n",
    "    \n",
    "  #Plot pseudotime in dimentional reduction\n",
    "    name.output=str_c(dir.name, name, sep='/') %>% str_c(\"from\", start, red.use, \"pseudotime.tiff\", sep = \".\")\n",
    "    ppi <- 600\n",
    "    tiff(name.output, width=3.8*ppi, height=3*ppi, res=ppi)\n",
    "    #Output twice size of figure, compress after pasting in the powerpoint.\n",
    "    p <- FeaturePlot(object = seu2, features = \"pseudotime\", cols = c(\"#800080\", \"#ffee00\"), reduction = red.use, pt.size = 0.1) +\n",
    "          theme(plot.title = element_blank(), \n",
    "                axis.title.x = element_text(size=16, family = \"Arial\"), \n",
    "                axis.title.y = element_text(size=16, family = \"Arial\"), \n",
    "                axis.text.x = element_text(size=12, colour = 1, family = \"Arial\"), \n",
    "                axis.text.y = element_text(size = 12, colour = 1, family = \"Arial\"),\n",
    "               legend.title = element_text(size=12, family = \"Arial\"),\n",
    "               legend.text = element_text(size=12, family = \"Arial\"),\n",
    "               legend.key.width = unit(0.5, \"cm\")) +\n",
    "          theme(panel.border = element_rect(fill = NA, size = 0.75)) \n",
    "    plot(p)\n",
    "    dev.off()\n",
    "    \n",
    "  return(seu2)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "60834307",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Append sample origin to meta.data\n",
    "seu@meta.data$sample.origin <- NA\n",
    "seu@meta.data$sample.origin[which(seu@meta.data$orig.ident == \"mouseSampleTag5\" | seu@meta.data$orig.ident == \"mouseSampleTag6\")] <- \"aCD4\"\n",
    "seu@meta.data$sample.origin[which(seu@meta.data$orig.ident == \"mouseSampleTag1\")] <- \"Control\"\n",
    "seu@meta.data$sample.origin[which(seu@meta.data$orig.ident == \"mouseSampleTag2\")] <- \"aPDL1\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "d271ef2d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "class: SlingshotDataSet \n",
      "\n",
      " Samples Dimensions\n",
      "   28098         26\n",
      "\n",
      "lineages: 2 \n",
      "Lineage1: 1  4  3  0  \n",
      "Lineage2: 1  4  2  \n",
      "\n",
      "curves: 2 \n",
      "Curve1: Length: 34.883\tSamples: 22429.03\n",
      "Curve2: Length: 28.966\tSamples: 15775.05\n"
     ]
    }
   ],
   "source": [
    "###Pseudotime analysis for subset of clusters\n",
    "sample.basename <- \"Pooled\"\n",
    "dir.name <- \"results/preprocess/Pseudotime\"\n",
    "start <- \"1\" #starting clusters\n",
    "subcluster <- c(0,1,2,3,4)\n",
    "red.use <- \"umap\"\n",
    "\n",
    "############################# Processing layer ######################################\n",
    "dir.create(dir.name, recursive = TRUE)\n",
    "\n",
    "#Call seurat object and condition name\n",
    "seu.sub <- subset(seu, idents = subcluster)\n",
    "\n",
    "#Define PCs for Pseudotime analysis (Call the result of Jackstraw)\n",
    "tmp = as.data.frame(seu@reductions$pca@jackstraw@overall.p.values)\n",
    "tmp1 = tmp[tmp$Score>0.05,1]\n",
    "dims= c(1:(min(tmp1)-1))\n",
    "\n",
    "#Perform pseudotime analysis\n",
    "seu <- Pseudotime.Min(seu, seu.sub, sample.name, dims, start, dir.name)\n",
    "\n",
    "#save seurat object with pseudotime\n",
    "file.name <- str_c(dir.seurat, sample.name, sep = \"/\") %>% str_c(\"_clust_pseudotime.rda\", sep='')\n",
    "save(seu, file=file.name)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee792974",
   "metadata": {},
   "source": [
    "### Add Gene Scores for Tex term, Tex prog, Cytotoxic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "6656855a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“The following features are not present in the object: Cd244, not searching for symbol synonyms”\n"
     ]
    }
   ],
   "source": [
    "# Introduce Gene Scores using AddModuleScore function\n",
    "\n",
    "#Put signature gene set into \"Signatures\" directory\n",
    "dir.input <- \"Signatures\"\n",
    "sample.name <- \"Pooled\"\n",
    "\n",
    "########################## Processing layer #############################\n",
    "#load Signature data\n",
    "files  <- list.files(dir.input, pattern=\"Sig.\")\n",
    "\n",
    "for(i in files){\n",
    "      i <- str_c(dir.input, i, sep = \"/\")\n",
    "      gene_list <- read.table(i, header = TRUE)\n",
    "      gene_list <- list(as.vector(gene_list$Genes))\n",
    "      sig_name <- str_remove(i, dir.input)\n",
    "      sig_name <- str_remove(sig_name, \"/Sig.\")\n",
    "      sig_name <- str_remove(sig_name, \".txt\")\n",
    "      seu <- AddModuleScore(object = seu, \n",
    "                            ctrl=10,\n",
    "                            features = gene_list,\n",
    "                            name=sig_name)\n",
    "    }\n",
    "#Output Seurat object\n",
    "file.name <- str_c(dir.seurat, sample.name, sep = \"/\") %>% str_c(\"_clust_pseudotime_Signature.rda\", sep='')\n",
    "save(seu, file=file.name)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8865e2ad",
   "metadata": {},
   "source": [
    "### Integration of scTCRseq results into SCT Seurat object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "a3377919",
   "metadata": {},
   "outputs": [],
   "source": [
    "###Define function\n",
    "#Create histogram was showing the total number of TCR reads detected in each cell (top left) \n",
    "#and the percentage of TCR reads that were the most common among them (bottom left).\n",
    "#Scatter plot representing these two metrices of cells were also generated\n",
    "Histogram <- function(d, file.name, sample.name, dir.output){\n",
    "\n",
    "  d_count_total <- as.numeric(d$count.total)\n",
    "  \n",
    "  #Count histogram\n",
    "  ppi <- 300\n",
    "  image.file <- str_c(dir.output, file.name, sep = \"/\") %>% str_c(sample.name, \"histogram.count.tiff\", sep = '.') \n",
    "  tiff(image.file, width=1.2*ppi, height=0.8*ppi, res=ppi)\n",
    "  p <- ggplot(NULL, aes(x=d_count_total)) +\n",
    "    geom_histogram(binwidth=0.1) +\n",
    "    theme_bw(base_size = 6) +\n",
    "    labs(x = \"Read counts of TCR\") +\n",
    "    theme(\n",
    "      axis.title.y=element_blank(),\n",
    "      axis.text.x = element_text(family=\"Arial\"),\n",
    "      axis.text.y = element_text(family=\"Arial\"),\n",
    "      axis.title=element_text(size=4)) +\n",
    "    scale_x_continuous(trans=scales::log2_trans(),\n",
    "                   breaks=scales::trans_breaks(\"log2\",function(x) 2^x),\n",
    "                   labels=scales::trans_format(\"log2\",scales::math_format(2^.x)))\n",
    "  print(p)\n",
    "  dev.off()\n",
    "\n",
    "  #Frequency histogram  \n",
    "  d_freq <- as.numeric(d$freq)\n",
    "  med_f <- 0.3\n",
    "  ppi <- 300\n",
    "  image.file <- str_c(dir.output, file.name, sep = \"/\") %>% str_c(sample.name, \"histogram.freq.tiff\", sep = '.')\n",
    "  tiff(image.file, width=1.2*ppi, height=0.8*ppi, res=ppi)\n",
    "  p <- ggplot(NULL, aes(x=d_freq)) +\n",
    "    geom_histogram(binwidth=0.05) +\n",
    "    theme_bw(base_size = 6) +\n",
    "    labs(x = \"Proportion of the largest TCR read\") +\n",
    "    theme(\n",
    "      axis.title.y=element_blank(),\n",
    "      axis.text.x = element_text(family=\"Arial\"),\n",
    "      axis.text.y = element_text(family=\"Arial\"),\n",
    "      axis.title=element_text(size=4)) \n",
    "  print(p)\n",
    "  dev.off()\n",
    "\n",
    "  df <- data.frame(d_count_total, d_freq)\n",
    "  \n",
    "  #Scatter plot    \n",
    "  total <- nrow(subset(df, df$d_freq >= 0))\n",
    "  lb <- paste(round(100*nrow( subset(df, df$d_freq < 0.6 & df$d_count_total < 2^5 )) / total, digits = 1), \"%\", sep=\"\" ) \n",
    "  rb <- paste(round(100*nrow( subset(df, df$d_freq >= 0.6 & df$d_count_total < 2^5 )) / total, digits = 1), \"%\", sep=\"\" ) \n",
    "  lt <- paste(round(100*nrow( subset(df, df$d_freq < 0.6 & df$d_count_total >= 2^5 )) / total, digits = 1), \"%\", sep=\"\" ) \n",
    "  rt <- paste(round(100*nrow( subset(df, df$d_freq >= 0.6 & df$d_count_total >= 2^5 )) / total, digits = 1), \"%\", sep=\"\" ) \n",
    "  \n",
    "  ppi <- 300\n",
    "  image.file <- str_c(dir.output, file.name, sep = \"/\") %>% str_c(sample.name, \"Scatter.tiff\", sep = '.')                     \n",
    "  tiff(image.file, width=1.2*ppi, height=1.2*ppi, res=ppi)\n",
    "  p <- ggplot(d, aes(x=d_freq, y=d_count_total)) +  \n",
    "    stat_bin2d(bins=60) +\n",
    "    scale_fill_gradient(low=\"lightblue\", high=\"red\") +\n",
    "    theme_bw(base_size = 6) +\n",
    "    geom_vline(aes(xintercept = 0.6), size=0.25, colour=\"black\") +\n",
    "    geom_hline(aes(yintercept = 2^5), size=0.25, colour=\"black\") +\n",
    "    labs(x = \"Proportion of the largest TCR read\", y = \"Read counts of TCR\") +\n",
    "    theme(\n",
    "      axis.text.x = element_text(family=\"Arial\"),\n",
    "      axis.text.y = element_text(family=\"Arial\"),\n",
    "      axis.title=element_text(size=4)) +\n",
    "    guides(fill=FALSE) +\n",
    "    scale_y_continuous(trans=scales::log2_trans(),\n",
    "                   breaks=scales::trans_breaks(\"log2\",function(x) 2^x),\n",
    "                   labels=scales::trans_format(\"log2\",scales::math_format(2^.x)))\n",
    "    p <- p + #xlim(0,1) +\n",
    "    annotate(\"text\", x=-Inf, y=0, hjust=-0.1, vjust=-0.4, label=lb,\n",
    "             family=\"Arial\",colour=\"black\",size=1.5) +\n",
    "    annotate(\"text\", x=Inf, y=0, hjust=1.1, vjust=-0.4, label=rb,\n",
    "             family=\"Arial\",colour=\"black\",size=1.5) +\n",
    "    annotate(\"text\", x=-Inf, y=Inf, hjust=-0.1, vjust=1.3, label=lt,\n",
    "             family=\"Arial\",colour=\"black\",size=1.5) +\n",
    "    annotate(\"text\", x=Inf, y=Inf, hjust=1.1, vjust=1.3, label=rt,\n",
    "             family=\"Arial\",colour=\"black\",size=1.5) \n",
    "  print(p)\n",
    "  dev.off()  \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "219781ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "###Define function\n",
    "#Create TCRa/TCRb combine table for cell barcodes\n",
    "CombineTable <- function(dir.input, sample.name, dir.output, count_th, freq_th, clonotype, cores){\n",
    "    combined.tables <- data.frame()\n",
    "    #Unzip files\n",
    "    files <- list.files(dir.input, sample.name)\n",
    "    for(file in files){\n",
    "        name.input <- str_c(dir.input, file, sep = \"/\")\n",
    "        bunzip2(name.input, remove=FALSE)\n",
    "        name.input <- str_remove(name.input, pattern = \".bz2\")\n",
    "        untar(name.input)\n",
    "        file.remove(name.input)\n",
    "    }\n",
    "    \n",
    "    ###Define function\n",
    "    tableread_fast = function(i, header=TRUE, quote=\"\", sep=\",\"){\n",
    "      tmp = fread(i, header=header, sep=sep, quote=quote, nThread=32)\n",
    "      tmp = as.data.frame(tmp)\n",
    "      return(tmp)\n",
    "    }\n",
    "    \n",
    "    #Convert MiXCR output into VDJtools format\n",
    "    Convert <- function(file.name, name.TRA){\n",
    "        name.input <- str_c(name.TRA, file.name, sep = \"/\")\n",
    "        data <- tableread_fast(name.input, header = TRUE, sep = '\\t')\n",
    "\n",
    "        #Extract TCR information from mixcr output\n",
    "        count.total<-data$cloneCount / data$cloneFraction\n",
    "        freq<-data$cloneFraction\n",
    "        cdr3nt<-data$nSeqImputedCDR3\n",
    "        cdr3aa<-data$aaSeqImputedCDR3\n",
    "        v<-str_sub(data$bestVHit, end=-4)\n",
    "        d<-str_sub(data$bestDHit, end=-4)\n",
    "        j<-str_sub(data$bestJHit, end=-4)\n",
    "        data3 <- rbind(count.total,freq,cdr3nt,cdr3aa,v,d,j)\n",
    "        data3 <- as.data.frame(t(data3))\n",
    "        names(data3) <- c(\"count.total\",\"freq\",\"cdr3nt\",\"cdr3aa\",\"v\",\"d\",\"j\")\n",
    "\n",
    "        #Extract largest clone in cell barcode and append cell barcode information\n",
    "        name_out <- str_split(file.name, \"_\")\n",
    "        CB <- name_out[[1]][[4]]\n",
    "        CB_out <- str_split(CB, \"\\\\.\")\n",
    "        CB <- CB_out[[1]][[1]]\n",
    "        d_out <- data3[1,]\n",
    "        d_out$CB <- CB\n",
    "\n",
    "        return(d_out)\n",
    "    }\n",
    "\n",
    "    #Convert MiXCR output into VDJtools format\n",
    "    name.TRA <- str_c(sample.name, \"TRAC1_mixcr\", sep = \"_\")\n",
    "    files  <- list.files(name.TRA, pattern=\".txt\")\n",
    "    cl <- makeCluster(cores)\n",
    "    registerDoParallel(cl)\n",
    "    TRA.table <- invisible(foreach(file.name = files,\n",
    "            .combine = rbind, .packages=c(\"ggplot2\", \"extrafont\", \"stringr\", \"dplyr\", \"data.table\")) %dopar% {Convert(file.name, name.TRA)})\n",
    "    stopCluster(cl)\n",
    "    unlink(name.TRA, recursive=TRUE)\n",
    "\n",
    "    name.TRB <- str_c(sample.name, \"TRBC1_mixcr\", sep = \"_\")\n",
    "    files  <- list.files(name.TRB, pattern=\".txt\")\n",
    "    cl <- makeCluster(cores)\n",
    "    registerDoParallel(cl)\n",
    "    TRB.table <- invisible(foreach(file.name = files,\n",
    "            .combine = rbind, .packages=c(\"ggplot2\", \"extrafont\", \"stringr\", \"dplyr\", \"data.table\")) %dopar% {Convert(file.name, name.TRB)})\n",
    "    stopCluster(cl)\n",
    "    unlink(name.TRB, recursive=TRUE)\n",
    "\n",
    "    #Output histogram and scatter plot for summarizing scTCR status\n",
    "    Histogram(TRA.table, \"TRA\", sample.name, dir.output)\n",
    "    Histogram(TRB.table, \"TRB\", sample.name, dir.output)\n",
    "\n",
    "    #thresholding\n",
    "    TRA.table$count.total <- as.numeric(TRA.table$count.total)\n",
    "    TRA.table$freq <- as.numeric(TRA.table$freq)\n",
    "    TRB.table$count.total <- as.numeric(TRB.table$count.total)\n",
    "    TRB.table$freq <- as.numeric(TRB.table$freq)\n",
    "    TCRa_th <- subset(TRA.table, count.total >= 2^(count_th) & freq >= freq_th)\n",
    "    TCRb_th <- subset(TRB.table, count.total >= 2^(count_th) & freq >= freq_th)\n",
    "\n",
    "    #Paring TCRa and TCRb by cell barcode\n",
    "    names(TCRa_th) <- c(\"count.total.A\",\"freq.A\",\"cdr3nt.A\",\"cdr3aa.A\",\"v.A\",\"d.A\",\"j.A\", \"CB\")\n",
    "    names(TCRb_th) <- c(\"count.total.B\",\"freq.B\",\"cdr3nt.B\",\"cdr3aa.B\",\"v.B\",\"d.B\",\"j.B\", \"CB\")\n",
    "    combined <- merge(TCRa_th, TCRb_th, all=T, by =\"CB\")\n",
    "    combined$CB <- str_c(sample.name, combined$CB, sep = \"_\")\n",
    "    #Exclude cells in which neither TCRa nor TCRb sequence were detected\n",
    "    combined <- subset(combined, combined$cdr3nt.A != \"UD\" | combined$cdr3nt.B != \"UD\")\n",
    "\n",
    "    #Generate clone id for further analysis\n",
    "    #Definition of clones can be changed between \"ABnt\", \"ABaa\", \"Bnt\", and \"Baa\"\n",
    "    if(clonotype==\"nt\"){\n",
    "        combined$clone.id.TCRa <- str_c(combined$cdr3nt.A, combined$v.A, combined$j.A, sep=\"_\")\n",
    "        combined$clone.id.TCRb <- str_c(combined$cdr3nt.B, combined$v.B, combined$j.B, sep=\"_\")\n",
    "    }\n",
    "    if(clonotype==\"aa\"){\n",
    "        combined$clone.id.TCRa <- str_c(combined$cdr3aa.A, combined$v.A, combined$j.A, sep=\"_\")\n",
    "        combined$clone.id.TCRb <- str_c(combined$cdr3aa.B, combined$v.B, combined$j.B, sep=\"_\")\n",
    "    }\n",
    "    combined$clone.id.TCRab <- str_c(combined$clone.id.TCRa, combined$clone.id.TCRb, sep = \"_\")\n",
    "    name.output <- str_c(dir.output, sample.name, sep = \"/\") %>% str_c(clonotype, \"table\", \"count_th\", count_th, \"freq_th\", freq_th, clonotype, \"csv\", sep = \".\")\n",
    "    write.csv(combined, name.output, row.names = FALSE)\n",
    "    \n",
    "    combined.tables <- rbind(combined.tables, combined)\n",
    "    \n",
    "    return(combined.tables)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "d0203dff",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Removed 170 rows containing non-finite values (stat_bin).”\n",
      "Warning message:\n",
      "“Removed 170 rows containing non-finite values (stat_bin).”\n",
      "Warning message:\n",
      "“`guides(<scale> = FALSE)` is deprecated. Please use `guides(<scale> = \"none\")` instead.”\n",
      "Warning message:\n",
      "“Transformation introduced infinite values in continuous y-axis”\n",
      "Warning message:\n",
      "“Transformation introduced infinite values in continuous y-axis”\n",
      "Warning message:\n",
      "“Removed 170 rows containing non-finite values (stat_bin2d).”\n",
      "Warning message:\n",
      "“Removed 156 rows containing non-finite values (stat_bin).”\n",
      "Warning message:\n",
      "“Removed 156 rows containing non-finite values (stat_bin).”\n",
      "Warning message:\n",
      "“`guides(<scale> = FALSE)` is deprecated. Please use `guides(<scale> = \"none\")` instead.”\n",
      "Warning message:\n",
      "“Transformation introduced infinite values in continuous y-axis”\n",
      "Warning message:\n",
      "“Transformation introduced infinite values in continuous y-axis”\n",
      "Warning message:\n",
      "“Removed 156 rows containing non-finite values (stat_bin2d).”\n",
      "Warning message:\n",
      "“Removed 2721 rows containing non-finite values (stat_bin).”\n",
      "Warning message:\n",
      "“Removed 2721 rows containing non-finite values (stat_bin).”\n",
      "Warning message:\n",
      "“`guides(<scale> = FALSE)` is deprecated. Please use `guides(<scale> = \"none\")` instead.”\n",
      "Warning message:\n",
      "“Transformation introduced infinite values in continuous y-axis”\n",
      "Warning message:\n",
      "“Transformation introduced infinite values in continuous y-axis”\n",
      "Warning message:\n",
      "“Removed 2721 rows containing non-finite values (stat_bin2d).”\n",
      "Warning message:\n",
      "“Removed 1549 rows containing non-finite values (stat_bin).”\n",
      "Warning message:\n",
      "“Removed 1549 rows containing non-finite values (stat_bin).”\n",
      "Warning message:\n",
      "“`guides(<scale> = FALSE)` is deprecated. Please use `guides(<scale> = \"none\")` instead.”\n",
      "Warning message:\n",
      "“Transformation introduced infinite values in continuous y-axis”\n",
      "Warning message:\n",
      "“Transformation introduced infinite values in continuous y-axis”\n",
      "Warning message:\n",
      "“Removed 1549 rows containing non-finite values (stat_bin2d).”\n"
     ]
    }
   ],
   "source": [
    "#Extract TCR sequence for each cel barcode\n",
    "#iteration process for dataset (CD4/CD8 Seurat object) and clonotype (Ant/Bnt)\n",
    "set.seed(seed = 42)\n",
    "sample.list <- c(\"Ex124TCRBD\", \"ICIrevTCR1\")\n",
    "dir.input <- \"data/scTCR_rawdata\"\n",
    "cores <- 12\n",
    "dir.output <- \"results/preprocess/scTCR_processing\"\n",
    "#Threshold for valid cell barcode with single TCR sequence\n",
    "count_th <- 5 #read count threshold for all TCR sequences per cell barcode\n",
    "freq_th <- 0.6 #proportion threshold for largest TCR\n",
    "clonotype <- \"nt\" #Definition of clones. nt: nucleotide sequence / aa: amino acid sequnece \n",
    "\n",
    "################################ Processing layer #################################################\n",
    "dir.create(dir.output)\n",
    "\n",
    "combined.tables <- data.frame()\n",
    "\n",
    "#Create TCRa/TCRb paring table\n",
    "for(sample.name in sample.list){\n",
    "    #Unzip files\n",
    "    ##Create TCRa/TCRb combine table for cell barcodes\n",
    "    combined.table <- CombineTable(dir.input, sample.name, dir.output, count_th, freq_th, clonotype, cores)\n",
    "    combined.tables <- rbind(combined.tables, combined.table)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "82d5fe09",
   "metadata": {},
   "outputs": [],
   "source": [
    "###Integrate scTCRdata into the Seurat object\n",
    "sample.name <- \"Pooled\"\n",
    "\n",
    "################################ Processing layer #################################################\n",
    "#Change batch name in combine.table\n",
    "combined.tables$CB <- str_replace(combined.tables$CB, \"Ex124TCRBD\", \"Aoki\")\n",
    "combined.tables$CB <- str_replace(combined.tables$CB, \"TCR\", \"Target\")\n",
    "\n",
    "#Extract meta.data and cell BC information\n",
    "meta.data <- seu@meta.data\n",
    "meta.data$names <- row.names(meta.data)\n",
    "CB.info <- str_split(meta.data$names, pattern = \"_\", simplify = TRUE)\n",
    "meta.data$CB <- str_c(CB.info[,2], CB.info[,3], sep = \"_\")\n",
    "\n",
    "#Merge to Seurat object of SCT data\n",
    "meta.data <- merge(combined.tables, meta.data, all.y = T, by =\"CB\")\n",
    "clone.ids <- c(\"clone.id.TCRa\", \"clone.id.TCRb\", \"clone.id.TCRab\")\n",
    "for(clone.id in clone.ids){\n",
    "    clone.id.list <- dplyr::select(meta.data, clone.id)\n",
    "    row.names(clone.id.list)=as.character(meta.data$names)\n",
    "    seu <- AddMetaData(object = seu, metadata = clone.id.list, col.name = clone.id)\n",
    "}\n",
    "    \n",
    "#Output Seurat object\n",
    "file.name <- str_c(dir.seurat, sample.name, sep = \"/\") %>% str_c(\"_clust_pseudotime_Signature_scTCR.rda\", sep='')\n",
    "save(seu, file=file.name)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4b02b33",
   "metadata": {},
   "source": [
    "### SCT analysis after integrating scTCRseq results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "39c86934",
   "metadata": {},
   "outputs": [],
   "source": [
    "###Define function\n",
    "\n",
    "##Make clone table by assembling clones from scTCR results \n",
    "CloneSummary <- function(seu, dir.name, sample.name, clonotype){\n",
    "    #load metadata\n",
    "    meta.data <- as.data.frame(seu@meta.data)\n",
    "    meta.data$CellBC <- row.names(meta.data)\n",
    "    \n",
    "    #extract cells only in the assigned sample\n",
    "    meta.data <- dplyr::filter(meta.data, str_detect(sample.origin, sample.name))\n",
    "    \n",
    "    #Define clone by TCRa / TCRb/ TCRa&b\n",
    "    if(clonotype == \"TCRa\"){\n",
    "       meta.data$clone.id <- meta.data$clone.id.TCRa \n",
    "    }\n",
    "    if(clonotype == \"TCRb\"){\n",
    "       meta.data$clone.id <- meta.data$clone.id.TCRb \n",
    "    }\n",
    "    if(clonotype == \"TCRab\"){\n",
    "       meta.data$clone.id <- meta.data$clone.id.TCRab \n",
    "    } \n",
    "\n",
    "    #Summarize cluster distribution of each clone\n",
    "    tmp_out = table(meta.data$seurat_clusters, meta.data$clone.id)\n",
    "    tmp_out2 <- as.data.frame(tmp_out, row.names = NULL,\n",
    "                  responseName = \"Freq\", stringsAsFactors = TRUE,\n",
    "                  sep = \"\", base = list(LETTERS))\n",
    "    names(tmp_out2) <- c(\"Clust\", \"names\", \"Freq\")\n",
    "    tmp_out2 <- dcast(tmp_out2, names ~ Clust)\n",
    "\n",
    "    #Summarize the count and frequency of clones \n",
    "    tmp = meta.data %>% group_by(`clone.id`) %>%\n",
    "      dplyr::summarise(count = n()) %>%\n",
    "      dplyr::arrange(desc(count))\n",
    "    tmp = as.data.frame(tmp)\n",
    "    tmp <- tmp[!is.na(tmp$clone.id), ] #NAである細胞を除く\n",
    "    tmp[,3]=tmp[,2]/sum(tmp[,2])\n",
    "    colnames(tmp)=c(\"ntSeq_TRA_TRB_freq\", \"CloneCount\", \"CloneFreq\")\n",
    "    tmp <- tmp[order(tmp$ntSeq_TRA_TRB_freq),]\n",
    "    temp_count <- as.vector(tmp$CloneCount)\n",
    "    temp_freq <- as.vector(tmp$CloneFreq)\n",
    "    #Combine to the cluster distribution\n",
    "    tmp_out3 <- cbind(temp_count, temp_freq, tmp_out2)\n",
    "\n",
    "    #Assign ranks to clones\n",
    "    tmp_out3 <- tmp_out3[order(tmp_out3$temp_freq, decreasing=T),]\n",
    "    rank <-  1:nrow(tmp_out3)\n",
    "    tmp_out3 <- cbind(rank, tmp_out3)\n",
    "    tmp_out3 <- data.frame(tmp_out3)\n",
    "    tmp_out3$rank <- paste(\"Top\",tmp_out3$rank, sep=\"\")\n",
    "    \n",
    "    #Output\n",
    "    name.out <- str_c(dir.name, sample.name, sep = \"/\") %>% str_c(clonotype, \"clone_within_cluster.txt\", sep = \".\")\n",
    "    write.table(tmp_out3, name.out, row.names=F, col.names=T, sep=\"\\t\", quote=F) \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "8be21aaf",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using Freq as value column: use value.var to override.\n",
      "\n",
      "Using Freq as value column: use value.var to override.\n",
      "\n",
      "Using Freq as value column: use value.var to override.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Main module\n",
    "\n",
    "###Input layer \n",
    "sample.name <- \"Pooled\"\n",
    "dir.name <- \"results/preprocess/scTCR_analysis\"\n",
    "treats <- c(\"Control\", \"aPDL1\", \"aCD4\")\n",
    "clonotype <- \"nt\" #Definition of clones. nt: nucleotide sequence / aa: amino acid sequnece \n",
    "\n",
    "########################## Processing layer #############################\n",
    "dir.create(dir.name, recursive = TRUE)\n",
    "    \n",
    "#Make clone table by assembling clones from scTCR results \n",
    "for(treat in treats){\n",
    "    CloneSummary(seu, dir.name, treat, \"TCRb\") \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "28be195c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(dir.name):\n",
      "“'results/preprocess/scTCR_analysis' already exists”\n"
     ]
    }
   ],
   "source": [
    "### Detect dLN-Tumor OL clones\n",
    "\n",
    "###Input layer\n",
    "#Specify query files: frequency and cluster distribution table of scTCR clones\n",
    "dir.name <- \"results/preprocess/scTCR_analysis\"\n",
    "treats <- c(\"aCD4\", \"aPDL1\", \"Control\")\n",
    "#Specify the repertoire data of dLN\n",
    "dir.bulk <- \"data/BulkTCR_rawdata\"\n",
    "\n",
    "cores <- 4\n",
    "\n",
    "########################## Processing layer #############################\n",
    "dir.create(dir.name)\n",
    "    \n",
    "for(treat in treats){\n",
    "    #load scTCR clone table file\n",
    "    name.query <- str_c(dir.name, treat, sep = \"/\") %>% str_c(\"TCRb.clone_within_cluster.txt\", sep = \".\") \n",
    "    clone.q <- read.table(name.query, header = TRUE)\n",
    "    clone.q$TCRb <- clone.q$names\n",
    "\n",
    "    #load dLN reprtoire files\n",
    "    name.bulk <- str_c(dir.bulk, \"CD8_dLN_SCT_\", sep = \"/\") %>% str_c(treat, \".txt\", sep = \"\")\n",
    "    d <- tableread_fast(name.bulk, header = TRUE, sep=\"\\t\", quote = \"\\\"\")\n",
    "    #Reconstruct the information of clones for searching overlap\n",
    "    d$TCRb <- str_c(d$cdr3nt, d$v, d$j, sep = \"_\")\n",
    "    d <- dplyr::select(d, c(\"TCRb\", \"freq\"))\n",
    "    names(d) <- c(\"TCRb\", \"freq_dLN\")\n",
    "\n",
    "    #Search clones overlapped with Tumor scTCR repertoire\n",
    "    d_output <- merge(clone.q, d, by = \"TCRb\", all.x = T)\n",
    "    d_output$treat <- treat\n",
    "\n",
    "    ###Define Oligoclonal / Polyclonal fraction of dLN-tumor OL\n",
    "    d_output$OL <- \"nonOL\"\n",
    "    #Extract OL clones\n",
    "    data_LT <- dplyr::filter(d_output, !is.na(freq_dLN))\n",
    "    data_nonOL <- dplyr::filter(d_output, is.na(freq_dLN))\n",
    "    #Determine Oligoclonal / Polyclonal fraction\n",
    "    data_LT <- data_LT[order(data_LT$temp_freq, decreasing = T),]\n",
    "    count.poly <- nrow(data_LT)-10\n",
    "    data_LT$OL <- c(rep(\"Oligo\", times = 10), rep(\"Poly\", times = count.poly))\n",
    "    d_output <- rbind(data_LT, data_nonOL)\n",
    "        \n",
    "    ### Integrate dLN-tumor OL status (oligoclona OL / polyclonal OL / nonOL) and gene scores\n",
    "    ###Summarize Gene score for each TCR clone\n",
    "    meta.data <- as.data.frame(seu@meta.data)\n",
    "    meta.data.calc <- dplyr::filter(meta.data, sample.origin == treat) %>% dplyr::select(c(\"clone.id.TCRb\", \"TexKLR1\", \"Tumor.Prog1\", \"Tumor.Term1\"))\n",
    "    out_score <- aggregate(x=meta.data.calc[c(\"TexKLR1\", \"Tumor.Prog1\", \"Tumor.Term1\")],\n",
    "                         by=list(meta.data.calc$clone.id), FUN=mean)\n",
    "    #Integration\n",
    "    out <- merge(d_output, out_score, by.x = \"names\", by.y = \"Group.1\", all.x = T)\n",
    "\n",
    "    name.out <- str_c(dir.name, treat, sep = \"/\") %>% str_c(\"scTCR.dLNOL.csv\", sep = \".\")\n",
    "    write.csv(out, name.out, row.names = FALSE)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "873f556e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
