---
title: "Single cell TCRseq"
date: "`r Sys.Date()`"
author: "ImmunoGeneTeqs Inc."
output: 
    html_document:
      theme: united
      toc: true
      number_sections: true
      md_extensions: -ascii_identifiers
      css: "../styles_customized.css"

params:
  directory: "data"
  dir_TRA_mixcr: "TumorTCRab-AMPure_TRAC1_mixcr"
  dir_TRB_mixcr: "TumorTCRab-AMPure_TRBC1_mixcr"
  rda_input: "TumorTarget_res1"
  meta.data_input: "meta.data.csv"
  clonotype: "ABnt" #(ABnt, ABaa, Bnt, Baa)
---



```{r setup, echo=FALSE, cache=FALSE, message=FALSE, warning=FALSE}

suppressMessages(library(knitr))
suppressMessages(library(rmdformats))
suppressMessages(library(stringr))
suppressMessages(library(DT))
suppressMessages(library(doParallel))
suppressMessages(library(foreach))
suppressMessages(library(dplyr))
suppressMessages(library(data.table))
suppressMessages(library(shiny))
suppressMessages(library(ggplot2))
suppressMessages(library(extrafont))
suppressMessages(library(cowplot))
suppressMessages(library(magick))
suppressMessages(library(scales))
suppressMessages(library(Seurat))
suppressMessages(library(viridis))
suppressMessages(library(tidyverse))


custom_colors <- list()
colors_dutch <- c(
  '#FFC312','#C4E538','#12CBC4','#FDA7DF','#ED4C67',
  '#F79F1F','#A3CB38','#1289A7','#D980FA','#B53471',
  '#EE5A24','#009432','#0652DD','#9980FA','#833471',
  '#EA2027','#006266','#1B1464','#5758BB','#6F1E51'
)
colors_spanish <- c(
  '#40407a','#706fd3','#f7f1e3','#34ace0','#33d9b2',
  '#2c2c54','#474787','#aaa69d','#227093','#218c74',
  '#ff5252','#ff793f','#d1ccc0','#ffb142','#ffda79',
  '#b33939','#cd6133','#84817a','#cc8e35','#ccae62'
)
colors_custom = c('#a6cee3','#1f78b4','#b2df8a','#33a02c','#fb9a99','#e31a1c',
                      '#fdbf6f','#ff7f00','#cab2d6','#6a3d9a','#ffff99','#b15928',
                      '#49beaa', '#611c35', '#2708a0')
custom_colors$discrete <- unique(c(colors_dutch, colors_spanish, colors_custom))


## Global options
options(max.print="75")
opts_chunk$set(echo=FALSE,
	             cache=FALSE,
               prompt=FALSE,
               comment=NA,
               message=FALSE,
               warning=FALSE
               )
opts_knit$set(width=75)

```


```{r remove_files, echo=FALSE, cache=FALSE, message=FALSE, warning=FALSE}
dir_var = params$directory
dir_TRA = params$dir_TRA_mixcr
dir_TRB = params$dir_TRB_mixcr
rda_input = params$rda_input
meta.data_input = params$meta.data_input
clonotype = params$clonotype
dir.output = "Output.files"
dir.create("Others")


setwd( paste("..", dir_var, sep = "/") )
unlink("meta.data.afterscTCR*", recursive = FALSE)
unlink("*scTCRmerged*", recursive = FALSE)
unlink("filtered2", recursive = TRUE)
unlink("scTCR_analysis", recursive = TRUE)
unlink("figures.*", recursive = TRUE)

```

```{r remove_comma1, echo=FALSE, cache=FALSE, message=FALSE}


setwd( paste("..", dir_var, sep = "/") )

command <- sprintf(paste('find ./ -type f -name "*.tar.bz2" -exec tar -jxvf {} \\;', sep=""))
system(command)

delfiles <- dir(path="./" ,pattern="*.tar.bz2")
invisible(file.remove(file.path("./", delfiles)))

setwd( paste(".", dir_TRA, sep = "/") )

cores <- 12
files  <- list.files(pattern=".txt")

dir.create("filtered2")

tableread_fast = function(i, header=TRUE, sep="\t"){
  tmp = fread(i, header=header, sep=sep, quote="", nThread=32)
  tmp = as.data.frame(tmp)
  return(tmp)
}



#====================================================================

Scatter <- function(file.name){
  data <- tableread_fast(file.name, header = TRUE, sep = '\t')
  name <- basename(file.name)
  name_out <- str_c("filtered2", name, sep = "/")
  
  data$targetSequences <- gsub(",", "", data$targetSequences)
  data$allVAlignments <- gsub(",", "", data$allVAlignments)
  data$allDAlignments <- gsub(",", "", data$allDAlignments)
  data$allJAlignments <- gsub(",", "", data$allJAlignments)
  data$allCAlignments <- gsub(",", "", data$allCAlignments)
  data3 <- data[, colnames(data) != "nSeqImputedVDJTranscriptWithout5UTR"]
  data3[is.na(data3)] <- ""
  fwrite(data3, name_out, row.names = FALSE, sep="\t", quote=F)
}

#===============================================================

#t<-proc.time()
cl <- makeCluster(cores)
registerDoParallel(cl)
invisible(foreach(file.name = files,
        .combine = rbind, .packages=c("ggplot2", "extrafont", "stringr", "dplyr", "data.table")) %dopar% {Scatter(file.name)})
stopCluster(cl)
#proc.time()-t

```

```{r remove_comma2, echo=FALSE, cache=FALSE, message=FALSE}
setwd( paste("..", dir_var, dir_TRB, sep = "/") )

cores <- 12
files  <- list.files(pattern=".txt")

dir.create("filtered2")

tableread_fast = function(i, header=TRUE, sep="\t"){
  tmp = fread(i, header=header, sep=sep, quote="", nThread=32)
  tmp = as.data.frame(tmp)
  return(tmp)
}

#====================================================================

Scatter <- function(file.name){
  data <- tableread_fast(file.name, header = TRUE, sep = '\t')
  name <- basename(file.name)
  name_out <- str_c("filtered2", name, sep = "/")
  
  data$targetSequences <- gsub(",", "", data$targetSequences)
  data$allVAlignments <- gsub(",", "", data$allVAlignments)
  data$allDAlignments <- gsub(",", "", data$allDAlignments)
  data$allJAlignments <- gsub(",", "", data$allJAlignments)
  data$allCAlignments <- gsub(",", "", data$allCAlignments)
  data3 <- data[, colnames(data) != "nSeqImputedVDJTranscriptWithout5UTR"]
    data3[is.na(data3)] <- ""
    fwrite(data3, name_out, row.names = FALSE, sep="\t", quote=F)
}

#===============================================================

#t<-proc.time()
cl <- makeCluster(cores)
registerDoParallel(cl)
invisible(foreach(file.name = files,
        .combine = rbind, .packages=c("ggplot2", "extrafont", "stringr", "dplyr", "data.table")) %dopar% {Scatter(file.name)})
stopCluster(cl)
#proc.time()-t

```

```{r format_conversion1, echo=FALSE, cache=FALSE, message=FALSE}

dir.create(paste("..", dir_var, "filtered2", sep = "/"))
dir.create(paste("..", dir_var, "filtered2", "TRA", sep = "/"))
dir.create(paste("..", dir_var, "filtered2", "TRB", sep = "/"))

##TRA
path1<- paste("..", dir_var, dir_TRA, "filtered2", sep = "/") 
path2<- paste("..", dir_var, "filtered2", "TRA", sep = "/")

files = list.files(path1, pattern=".txt")
cores=12

tableread_fast = function(i, header=TRUE, sep="\t"){
  tmp = fread(i, header=header, sep=sep, quote="", nThread=32)
  tmp = as.data.frame(tmp)
  return(tmp)
}

#====================================================================

Convert <- function(file.name){
  data <- tableread_fast(paste(path1,file.name,sep="/"), header = TRUE, sep = '\t')
  name <- basename(file.name)
  name_out <- str_c(path2, name, sep = "/")
  
  count<-data$cloneCount
  freq<-data$cloneFraction
  ccr3nt<-data$nSeqImputedCDR3
  ccr3aa<-data$aaSeqImputedCDR3
  v<-gsub("[*00]", "", data$bestVHit)
  d<-gsub("[*00]", "", data$bestDHit)
  j<-gsub("[*00]", "", data$bestJHit)

  VEnd<-data$allVAlignments
  VEnd_0<-data$allVAlignments
  VEnd_1 <- str_split(VEnd_0, pattern = "[|]", simplify = TRUE)
  
  if(length(VEnd_1)==0){ #これは何も入ってない場合．無視したい
  } else if (is.na(VEnd_1)[1] == F) {                 
    VEnd <- as.numeric(VEnd_1[,5])-1 
  } else{
    VEnd<-data$allVAlignments
  }
  
  DStart<-data$allDAlignments
  DStart_0<-data$allDAlignments
  DStart_1 <- str_split(DStart_0, pattern = "[|]", simplify = TRUE)
  if(length(DStart_1)==0){ #これは何も入ってない場合．無視したい
  } else if (is.na(DStart_1)[1] == F) {                 
    DStart <- as.numeric(DStart_1[,4])      
  } else{
    DStart<-data$allDAlignments
  }
  
  DEnd<-data$allDAlignments
  DEnd_0<-data$allDAlignments
  DEnd_1 <- str_split(DEnd_0, pattern = "[|]", simplify = TRUE)
  
  if(length(DEnd_1)==0){ #これは何も入ってない場合．無視したい
  } else if (is.na(DEnd_1)[1] == F) {                 
    DEnd <- as.numeric(DEnd_1[,5])-1      
  } else{
    DEnd<-data$allDAlignments
  }
  
  JStart<-data$allJAlignments
  JStart_0<-data$allJAlignments
  JStart_1 <- str_split(JStart_0, pattern = "[|]", simplify = TRUE)
  
  if(length(JStart_1)==0){ #これは何も入ってない場合．無視したい
  } else if (is.na(JStart_1)[1] == F) {                 
    JStart <- as.numeric(JStart_1[,4])      
  } else{
    JStart <- data$allJAlignments
  }
  
  data3 <- rbind(count,freq,ccr3nt,ccr3aa,v,d,j,VEnd,DStart,DEnd,JStart)
  data3 <- as.data.frame(t(data3))
  names(data3) <- c("count","freq","ccr3nt","ccr3aa","v","d","j","VEnd","DStart","DEnd","JStart")
  
  fwrite(data3, name_out, col.names = T, sep="\t", quote=F)
}

cl <- makeCluster(cores)
registerDoParallel(cl)
invisible(foreach(file.name = files,
        .combine = rbind, .packages=c("ggplot2", "extrafont", "stringr", "dplyr", "data.table")) %dopar% {Convert(file.name)})
stopCluster(cl)

```


```{r format_conversion2, echo=FALSE, cache=FALSE, message=FALSE}

##TRB

path3<- paste("..", dir_var, dir_TRB, "filtered2", sep = "/") 
path4<- paste("..", dir_var, "filtered2", "TRB", sep = "/")

files = list.files(path3, pattern=".txt")
cores=12

tableread_fast = function(i, header=TRUE, sep="\t"){
  tmp = fread(i, header=header, sep=sep, quote="", nThread=32)
  tmp = as.data.frame(tmp)
  return(tmp)
}

#====================================================================

Convert <- function(file.name){
  data <- tableread_fast(paste(path3,file.name,sep="/"), header = TRUE, sep = '\t')
  name <- basename(file.name)
  name_out <- str_c(path4, name, sep = "/")
  
  count<-data$cloneCount
  freq<-data$cloneFraction
  ccr3nt<-data$nSeqImputedCDR3
  ccr3aa<-data$aaSeqImputedCDR3
  v<-gsub("[*00]", "", data$bestVHit)
  d<-gsub("[*00]", "", data$bestDHit)
  #if(length(d)==0){is.na(d)<-c(1)}
  j<-gsub("[*00]", "", data$bestJHit)
  
  VEnd<-data$allVAlignments  
  VEnd_0<-data$allVAlignments
  VEnd_1 <- str_split(VEnd_0, pattern = "[|]", simplify = TRUE)
  
  if(length(VEnd_1)==0){ #これは何も入ってない場合．無視したい
  } else if (is.na(VEnd_1)[1] == F) {                 
    VEnd <- as.numeric(VEnd_1[,5])-1 
  } else{
    VEnd<-data$allVAlignments
  }
  
  DStart<-data$allDAlignments
  DStart_0<-data$allDAlignments
  DStart_1 <- str_split(DStart_0, pattern = "[|]", simplify = TRUE)
  if(length(DStart_1)==0){ #これは何も入ってない場合．無視したい
  } else if (is.na(DStart_1)[1] == F) {                 
    DStart <- as.numeric(DStart_1[,4])      
  } else{
    DStart<-data$allDAlignments
  }
  
  DEnd<-data$allDAlignments
  DEnd_0<-data$allDAlignments
  DEnd_1 <- str_split(DEnd_0, pattern = "[|]", simplify = TRUE)
  #DEnd <- as.numeric(DEnd_1[,5])-1
  #DEnd <- ""
  
  if(length(DEnd_1)==0){ #これは何も入ってない場合．無視したい
  } else if (is.na(DEnd_1)[1] == F) {                 
    DEnd <- as.numeric(DEnd_1[,5])-1      
  } else{
    DEnd<-data$allDAlignments
  }
  
  JStart<-data$allJAlignments
  JStart_0<-data$allJAlignments
  JStart_1 <- str_split(JStart_0, pattern = "[|]", simplify = TRUE)
  
  if(length(JStart_1)==0){ #これは何も入ってない場合．無視したい
  } else if (is.na(JStart_1)[1] == F) {                 
    JStart <- as.numeric(JStart_1[,4])      
  } else{
    JStart <- data$allJAlignments
  }
  
  data3 <- rbind(count,freq,ccr3nt,ccr3aa,v,d,j,VEnd,DStart,DEnd,JStart)
  data3 <- as.data.frame(t(data3))
  names(data3) <- c("count","freq","ccr3nt","ccr3aa","v","d","j","VEnd","DStart","DEnd","JStart")
  
  fwrite(data3, name_out, col.names = T, sep="\t", quote=F)
}

cl <- makeCluster(cores)
registerDoParallel(cl)
invisible(foreach(file.name = files,
        .combine = rbind, .packages=c("ggplot2", "extrafont", "stringr", "dplyr", "data.table")) %dopar% {Convert(file.name)})
stopCluster(cl)

```



```{r combine_CB_result_nt1, echo=FALSE, cache=FALSE, message=FALSE}
#クローンテーブルをvdjtools formatへ変換したデータをInputに使用 (aaVJ)
#各Top 1 cloneを抽出してCell barcodeを付加し,データフレームにまとめる

setwd( paste("..", dir_var, "filtered2", "TRA", sep = "/") )

allele <- "TRA"   #ディレクトリに応じて変更

files  <- list.files(pattern=paste(allele,"txt", sep="."))
#files  <- list.files(pattern=allele)
cores <- 12

tableread_fast = function(i, header=TRUE, sep="\t"){
  tmp = fread(i, header=header, sep=sep, quote="", nThread=32)
  tmp = as.data.frame(tmp)
  return(tmp)
}

#====================================================================

Basic <- function(file.name){ 
  d <- tableread_fast(file.name, header = TRUE, sep = '\t')
  na <- basename(file.name)
  name_out <- str_split(na, "_")
  CB <- name_out[[1]][[4]]
  CB_out <- str_split(CB, "\\.")
  CB <- CB_out[[1]][[1]]
  
  d_out <- d[1,]
  d_out$CB <- CB
  
  return(d_out)
}

#===============================================================

#t<-proc.time()
cl <- makeCluster(cores)
registerDoParallel(cl)   
out <- foreach(file.name = files, .combine = rbind,
               .packages=c("stringr", "data.table")) %dopar% {Basic(file.name)}
stopCluster(cl)
#proc.time()-t

out <- as.data.frame(out)
names(out) <- c("count",	"freq",	"cdr3nt",	"cdr3aa",	"v",	"d",	"j",	"VEnd",	"DStart",	"DEnd",	"JStart", "CB")

w<-which(out$d == "")
is.na(out$d)<- w
out.name <- paste(allele, "CB.txt", sep="_")
write.table(out, out.name, row.names = F, sep="\t", quote = F)
```


```{r combine_CB_result_nt2, echo=FALSE, cache=FALSE, message=FALSE}
#クローンテーブルをvdjtools formatへ変換したデータをInputに使用 (aaVJ)
#各Top 1 cloneを抽出してCell barcodeを付加し,データフレームにまとめる

setwd( paste("..", dir_var, "filtered2", "TRB", sep = "/") )

allele <- "TRB"   #ディレクトリに応じて変更

files  <- list.files(pattern=paste(allele,"txt", sep="."))
#files  <- list.files(pattern=allele)
cores <- 12

tableread_fast = function(i, header=TRUE, sep="\t"){
  tmp = fread(i, header=header, sep=sep, quote="", nThread=32)
  tmp = as.data.frame(tmp)
  return(tmp)
}

#====================================================================

Basic <- function(file.name){ 
  d <- tableread_fast(file.name, header = TRUE, sep = '\t')
  na <- basename(file.name)
  name_out <- str_split(na, "_")
  CB <- name_out[[1]][[4]]
  CB_out <- str_split(CB, "\\.")
  CB <- CB_out[[1]][[1]]
  
  d_out <- d[1,]
  d_out$CB <- CB
  
  return(d_out)
}

#===============================================================

cl <- makeCluster(cores)
registerDoParallel(cl)   
out <- foreach(file.name = files, .combine = rbind,
               .packages=c("stringr", "data.table")) %dopar% {Basic(file.name)}
stopCluster(cl)

out <- as.data.frame(out)
names(out) <- c("count",	"freq",	"cdr3nt",	"cdr3aa",	"v",	"d",	"j",	"VEnd",	"DStart",	"DEnd",	"JStart", "CB")

w<-which(out$d == "")
is.na(out$d)<- w
out.name <- paste(allele, "CB.txt", sep="_")
write.table(out, out.name, row.names = F, sep="\t", quote = F)
```


# Single cell TCRseq結果のクオリティフィルタリング

各細胞で検出された総TCRリード数（左上）、およびその中で最多であったTCRリードの割合（左下）を示したヒストグラムを作成した。これから、十分な数のTCRリードが得られていない細胞、および他の細胞からのリーク等により十分な純度が得られていない細胞が少ないことを確認した。
続いて、上記2指標を用いて各細胞を散布図にプロットした（右）。各指標について一定の閾値を上回った細胞（散布図の右上の領域）のみを以降の解析に用いた。


--------

## TCRa

```{r histgram1, echo=FALSE, cache=FALSE, message=FALSE}

setwd( paste("..", dir_var, "filtered2", "TRA", sep = "/") )
cores <- 12
files  <- list.files(pattern="CB.txt")

#====================================================================
Histogram <- function(file.name){
  d <- read.table(file.name, header = TRUE)
  name <- basename(file.name)
  name <- str_replace(name, ".txt", "")

  d_count_total <- as.numeric(d$count) / as.numeric(d$freq)
  d_count_log <- d_count_total #log2(d_count_total)
  
  med <- log2(4)
  
  #Count, histogram
  ppi <- 300
  image.file <- paste('histogram', "count", name, 'tiff', sep = '.') 
  tiff(image.file, width=1.2*ppi, height=0.8*ppi, res=ppi)
  p <- ggplot(NULL, aes(x=d_count_log)) +
    geom_histogram(binwidth=0.1) +
#    geom_vline(aes(xintercept = med), size=0.25, colour="black") +
#    geom_vline(aes(xintercept = med+2), size=0.25, colour="red") +
#    geom_vline(aes(xintercept = med+4), size=0.25, colour="red") +
#    geom_vline(aes(xintercept = med+6), size=0.25, colour="blue") +
    theme_bw(base_size = 6) +
    labs(x = "Read counts of TCR") +
    theme(
      #axis.title.x=element_blank(), 
      axis.title.y=element_blank(),
      axis.text.x = element_text(family="Arial"),
      axis.text.y = element_text(family="Arial"),
      axis.title=element_text(size=4)) +
    scale_x_continuous(trans=scales::log2_trans(),
                   breaks=scales::trans_breaks("log2",function(x) 2^x),
                   labels=scales::trans_format("log2",scales::math_format(2^.x)))
  print(p)
  dev.off()

  #Frequency, histogram  
  d_freq <- as.numeric(d$freq)
  med_f <- 0.3
  ppi <- 300
  image.file <- paste('histogram', "freq", name, 'tiff', sep = '.') 
  tiff(image.file, width=1.2*ppi, height=0.8*ppi, res=ppi)
  p <- ggplot(NULL, aes(x=d_freq)) +
    geom_histogram(binwidth=0.05) +
#    geom_vline(aes(xintercept = med_f), size=0.25, colour="black") +
#    geom_vline(aes(xintercept = med_f+0.2), size=0.25, colour="red") +
#    geom_vline(aes(xintercept = med_f+0.4), size=0.25, colour="red") +
#    geom_vline(aes(xintercept = med_f+0.6), size=0.25, colour="blue") +
    theme_bw(base_size = 6) +
    labs(x = "Proportion of the largest TCR read") +
    theme(
      #axis.title.x=element_blank(), 
      axis.title.y=element_blank(),
      axis.text.x = element_text(family="Arial"),
      axis.text.y = element_text(family="Arial"),
      axis.title=element_text(size=4)) 
  print(p)
  dev.off()

  df <- data.frame(d_count_log, d_freq)
  
  #Scatter plot    
  total <- nrow(subset(df, df$d_freq >= 0))
  lb <- paste(round(100*nrow( subset(df, df$d_freq < 0.6 & df$d_count_log < 2^5 )) / total, digits = 1), "%", sep="" ) 
  rb <- paste(round(100*nrow( subset(df, df$d_freq >= 0.6 & df$d_count_log < 2^5 )) / total, digits = 1), "%", sep="" ) 
  lt <- paste(round(100*nrow( subset(df, df$d_freq < 0.6 & df$d_count_log >= 2^5 )) / total, digits = 1), "%", sep="" ) 
  rt <- paste(round(100*nrow( subset(df, df$d_freq >= 0.6 & df$d_count_log >= 2^5 )) / total, digits = 1), "%", sep="" ) 
  
  ppi <- 300
  image.file <- paste("Scatter", name, 'tiff', sep = '.') 
  tiff(image.file, width=1.2*ppi, height=1.2*ppi, res=ppi)
  p <- ggplot(d, aes(x=d_freq, y=d_count_log)) +  
    stat_bin2d(bins=30) +
    scale_fill_gradient(low="lightblue", high="red") +
    theme_bw(base_size = 6) +
    geom_vline(aes(xintercept = 0.6), size=0.25, colour="black") +
    geom_hline(aes(yintercept = 2^5), size=0.25, colour="black") +
    labs(x = "Proportion of the largest TCR read", y = "Read counts of TCR") +
    theme(
      #axis.title.x=element_blank(), axis.title.y=element_blank(),
      axis.text.x = element_text(family="Arial"),
      axis.text.y = element_text(family="Arial"),
      axis.title=element_text(size=4)) +
    guides(fill=FALSE) +
    scale_y_continuous(trans=scales::log2_trans(),
                   breaks=scales::trans_breaks("log2",function(x) 2^x),
                   labels=scales::trans_format("log2",scales::math_format(2^.x)))
    p <- p + #xlim(0,1) +
    annotate("text", x=-Inf, y=0, hjust=-0.1, vjust=-0.4, label=lb,
             family="Arial",colour="black",size=1.5) +
    annotate("text", x=Inf, y=0, hjust=1.1, vjust=-0.4, label=rb,
             family="Arial",colour="black",size=1.5) +
    annotate("text", x=-Inf, y=Inf, hjust=-0.1, vjust=1.3, label=lt,
             family="Arial",colour="black",size=1.5) +
    annotate("text", x=Inf, y=Inf, hjust=1.1, vjust=1.3, label=rt,
             family="Arial",colour="black",size=1.5)
  
  print(p)
  dev.off()  
  
}

#========================================================================

cl <- makeCluster(cores)
registerDoParallel(cl)   
invisible(foreach(file.name = files, .combine = rbind,
        .packages=c("ggplot2", "stringr")) %dopar% {Histogram(file.name)})
stopCluster(cl)


name = "TRA_CB"
p1 <- ggdraw() + draw_image(paste('histogram', "count", name, 'tiff', sep = '.'))
p2 <- ggdraw() + draw_image(paste('histogram', "freq", name, 'tiff', sep = '.'))
p3 <- ggdraw() + draw_image(paste("Scatter", name, 'tiff', sep = '.'))

#plot_grid(p1, p2, p3, nrow = 1) # labels = c(1,2,3))

cow_BC <- plot_grid(NULL, p1, NULL, p2,
                    # A negative rel_height shrinks space between elements
                    rel_heights = c(0.1, 1, -0.15, 1),
                    ncol = 1,
                    label_y = 1.07,
                    #labels = c("","count","","freq"),
                    label_size = 18,
                    label_fontfamily = "sans")

cow_D <- plot_grid(NULL, p3,
                    rel_heights = c(0.1, 1.85),
                    ncol = 1,
                    label_y = 1 + 0.07 /1.85,
                    #labels = c("","Scatter"),
                    label_size = 18,
                    label_fontfamily = "sans")


cow_BCD <- plot_grid(cow_BC, cow_D,
                     rel_widths = c(2,3),
                     scale=0.9, 
                     nrow = 1)

cow_BCD 


```

--------

## TCRb

```{r histgram2, echo=FALSE, cache=FALSE, message=TRUE}

setwd( paste("..", dir_var, "filtered2", "TRB", sep = "/") )

cores <- 12
files  <- list.files(pattern="CB.txt")

#====================================================================
Histogram <- function(file.name){
  d <- read.table(file.name, header = TRUE)
  name <- basename(file.name)
  name <- str_replace(name, ".txt", "")

  d_count_total <- as.numeric(d$count) / as.numeric(d$freq)
  d_count_log <- d_count_total
  
  med <- log2(4)
  
  #Count, histogram
  ppi <- 300
  image.file <- paste('histogram', "count", name, 'tiff', sep = '.') 
  tiff(image.file, width=1.2*ppi, height=0.8*ppi, res=ppi)
  p <- ggplot(NULL, aes(x=d_count_log)) +
    geom_histogram(binwidth=0.1) +
#    geom_vline(aes(xintercept = med), size=0.25, colour="black") +
#    geom_vline(aes(xintercept = med+2), size=0.25, colour="red") +
#    geom_vline(aes(xintercept = med+4), size=0.25, colour="red") +
#    geom_vline(aes(xintercept = med+6), size=0.25, colour="blue") +
    theme_bw(base_size = 6) +
    labs(x = "Read counts of TCR") +
    theme(
      #axis.title.x=element_blank(), 
      axis.title.y=element_blank(),
      axis.text.x = element_text(family="Arial"),
      axis.text.y = element_text(family="Arial"),
      axis.title=element_text(size=4)) +
    scale_x_continuous(trans=scales::log2_trans(),
                   breaks=scales::trans_breaks("log2",function(x) 2^x),
                   labels=scales::trans_format("log2",scales::math_format(2^.x)))
  print(p)
  dev.off()
  

  #Frequency, histogram  
  d_freq <- as.numeric(d$freq)
  med_f <- 0.3
  ppi <- 300
  image.file <- paste('histogram', "freq", name, 'tiff', sep = '.') 
  tiff(image.file, width=1.2*ppi, height=0.8*ppi, res=ppi)
  p <- ggplot(NULL, aes(x=d_freq)) +
    geom_histogram(binwidth=0.05) +
#    geom_vline(aes(xintercept = med_f), size=0.25, colour="black") +
#    geom_vline(aes(xintercept = med_f+0.2), size=0.25, colour="red") +
#    geom_vline(aes(xintercept = med_f+0.4), size=0.25, colour="red") +
#    geom_vline(aes(xintercept = med_f+0.6), size=0.25, colour="blue") +
    theme_bw(base_size = 6) +
    labs(x = "Proportion of the largest TCR read") +
    theme(
      #axis.title.x=element_blank(), 
      axis.title.y=element_blank(),
      axis.text.x = element_text(family="Arial"),
      axis.text.y = element_text(family="Arial"),
      axis.title=element_text(size=4)) 
  print(p)
  dev.off()

  df <- data.frame(d_count_log, d_freq)
  
  #Scatter plot
  total <- nrow(subset(df, df$d_freq >= 0))
  lb <- paste(round(100*nrow( subset(df, df$d_freq < 0.6 & df$d_count_log < 2^5 )) / total, digits = 1), "%", sep="" ) 
  rb <- paste(round(100*nrow( subset(df, df$d_freq >= 0.6 & df$d_count_log < 2^5 )) / total, digits = 1), "%", sep="" ) 
  lt <- paste(round(100*nrow( subset(df, df$d_freq < 0.6 & df$d_count_log >= 2^5 )) / total, digits = 1), "%", sep="" ) 
  rt <- paste(round(100*nrow( subset(df, df$d_freq >= 0.6 & df$d_count_log >= 2^5 )) / total, digits = 1), "%", sep="" ) 
  
  ppi <- 300
  image.file <- paste("Scatter", name, 'tiff', sep = '.') 
  tiff(image.file, width=1.2*ppi, height=1.2*ppi, res=ppi)
  p <- ggplot(d, aes(x=d_freq, y=d_count_log)) +  
    stat_bin2d(bins=30) +
    scale_fill_gradient(low="lightblue", high="red") +
    theme_bw(base_size = 6) +
    geom_vline(aes(xintercept = 0.6), size=0.25, colour="black") +
    geom_hline(aes(yintercept = 2^5), size=0.25, colour="black") +
    labs(x = "Proportion of the largest TCR read", y = "Read counts of TCR") +
    theme(
      #axis.title.x=element_blank(), axis.title.y=element_blank(),
      axis.text.x = element_text(family="Arial"),
      axis.text.y = element_text(family="Arial"),
      axis.title=element_text(size=4)) +
    guides(fill=FALSE) +
    scale_y_continuous(trans=scales::log2_trans(),
                   breaks=scales::trans_breaks("log2",function(x) 2^x),
                   labels=scales::trans_format("log2",scales::math_format(2^.x)))
    p <- p + #xlim(0,1) +
    annotate("text", x=-Inf, y=0, hjust=-0.1, vjust=-0.4, label=lb,
             family="Arial",colour="black",size=1.5) +
    annotate("text", x=Inf, y=0, hjust=1.1, vjust=-0.4, label=rb,
             family="Arial",colour="black",size=1.5) +
    annotate("text", x=-Inf, y=Inf, hjust=-0.1, vjust=1.3, label=lt,
             family="Arial",colour="black",size=1.5) +
    annotate("text", x=Inf, y=Inf, hjust=1.1, vjust=1.3, label=rt,
             family="Arial",colour="black",size=1.5)
  
  print(p)
  dev.off()  
  
}

#========================================================================

cl <- makeCluster(cores)
registerDoParallel(cl)   
invisible(foreach(file.name = files, .combine = rbind,
        .packages=c("ggplot2", "stringr")) %dopar% {Histogram(file.name)})
stopCluster(cl)



name = "TRB_CB"
p1 <- ggdraw() + draw_image(paste('histogram', "count", name, 'tiff', sep = '.'))
p2 <- ggdraw() + draw_image(paste('histogram', "freq", name, 'tiff', sep = '.'))
p3 <- ggdraw() + draw_image(paste("Scatter", name, 'tiff', sep = '.'))

#plot_grid(p1, p2, p3, nrow = 1) #labels = c(1,2,3))

cow_BC <- plot_grid(NULL, p1, NULL, p2,
                    # A negative rel_height shrinks space between elements
                    rel_heights = c(0.1, 1, -0.15, 1),
                    ncol = 1,
                    label_y = 1.07,
                    #labels = c("","count","","freq"),
                    label_size = 18,
                    label_fontfamily = "sans")

cow_D <- plot_grid(NULL, p3,
                    rel_heights = c(0.1, 1.85),
                    ncol = 1,
                    label_y = 1 + 0.07 /1.85,
                    #labels = c("","Scatter"),
                    label_size = 18,
                    label_fontfamily = "sans")


cow_BCD <- plot_grid(cow_BC, cow_D,
                     rel_widths = c(2,3),
                     scale=0.9, 
                      nrow = 1)

cow_BCD 
```

1.1 の棒グラフとヒストグラムについては，"/Output.files/TRA/" に .tiffファイルとして，それぞれ保存した．
また，1.2 の棒グラフとヒストグラムについては，"/Output.files/TRB/" に .tiffファイルとして，それぞれ保存した．


```{r paring, echo=FALSE, cache=FALSE, message=FALSE}

setwd( paste("..", dir_var, "filtered2",sep = "/") )

##CBに基づくTCRa, TCRbのparing
##combine_CB_result.Rの出力をinputに使用

##サンプル名の入力
TCRa.n <- "TRA_CB"
TCRa.name <- paste(TCRa.n, "txt", sep=".")
TCRb.n <- "TRB_CB"
TCRb.name <- paste(TCRb.n, "txt", sep=".")

#関数の導入
createEmptyDf = function( nrow, ncol, colnames = c() ){
  data.frame( matrix( vector(), nrow, ncol, dimnames = list( c(), colnames ) ) )
}

##Paring thresholdの設定
count_th_ <- 5
count_th_array <- c(count_th_)
freq_th_array <- c(0.6)
freq_th_<-"60p"
freq_th_name_array <- c(freq_th_)

for(k in 1:length(count_th_array)){
  count_th <- count_th_array[k]
  for(j in 1:length(freq_th_array)){
  freq_th <- freq_th_array[j]
  freq_th_name <- freq_th_name_array[j]

##出力名の作成
name.out <- paste(TCRa.n, TCRb.n, "count_th", count_th, "freq_th", freq_th_name, "csv", sep=".")

#読み込み
TCRa <- read.table(paste("TRA",TCRa.name,sep="/"), header = TRUE)
TCRb <- read.table(paste("TRB",TCRb.name,sep="/"), header = TRUE)

TCRa$count_total <- as.numeric(TCRa$count) / as.numeric(TCRa$freq)
TCRb$count_total <- as.numeric(TCRb$count) / as.numeric(TCRb$freq)

#thresholding
TCRa_th <- subset(TCRa, count_total >= 2^(count_th) & freq >= freq_th)
TCRb_th <- subset(TCRb, count_total >= 2^(count_th) & freq >= freq_th)

#CB, 統合
cb_list <- c(TCRa_th$CB, TCRb_th$CB)
cb_unique <- unique(cb_list)

#TCR情報書き出し用のデータフレーム作成
combined_cb_table <- data.frame()

#count, freqの書き出し用dfの作成
combine_cf_table = createEmptyDf(length(cb_unique), 4, colnames = c("A_count_total", "A_freq", "B_count_total", "B_freq") )

for (i in 1:length(cb_unique)){
  cbc <- cb_unique[i] 
  tcra <- subset(TCRa_th, CB == cbc)
  tcrb <- subset(TCRb_th, CB == cbc)
  if(nrow(tcra) == 0){
    tcra_sel <- c("UD", "UD", "UD", "UD", "UD")
    A_count_total <- "NA"
    A_freq <- "NA"
  }else{
    tcra_sel <- tcra[,3:7]
    A_count_total <- tcra$count_total
    A_freq <- tcra$freq
  }
  if(nrow(tcrb) == 0){
    tcrb_sel <- c("UD", "UD", "UD", "UD", "UD")
    B_count_total <- "NA"
    B_freq <- "NA"
  }else{
    tcrb_sel <- tcrb[,3:7]
    B_count_total <- tcrb$count_total
    B_freq <- tcrb$freq
  }
  combine_cb <- c(tcra_sel, tcrb_sel, cbc)
  combine_cb <- as.data.frame(combine_cb)
  names(combine_cb) = c("A_cdr3nt", "A_cdr3aa", "A_v", "A_d", "A_j",
                     "B_cdr3nt", "B_cdr3aa", "B_v", "B_d", "B_j", "CB")
  combined_cb_table <- rbind(combined_cb_table, combine_cb)

  #count, freq?̏????o??
  combine_cf <- c(A_count_total, A_freq, B_count_total, B_freq)
  combine_cf_table[i,] <- combine_cf
}

#Output
combine_table <- cbind(combined_cb_table, combine_cf_table)
combine_table <- as.data.frame(combine_table)
write.csv(combine_table, name.out, row.names = FALSE)
  }
}
```

--------

# Single cell TCRseqの性能評価

Single cell RNAseqでT細胞と同定された細胞数 (count.TB) 、TCRリードが検出された細胞数 (count.total)、およびTCRaとTCRbの両方が検出された細胞数 (count.paired) を示す。TCRaとTCRbのペアリング効率 (paired.ratio; count.paired/count.TB) も併せて示す。


```{r filtering, echo=FALSE, cache=FALSE, message=FALSE}
#"combine_CB_result.R"の出力結果を入力
#SCTによりEndo T cellと同定されたCBのみを抽出
#TCRa/TCRb Paring効率を算出

setwd( paste("..", dir_var, sep = "/") )

###Endo / PmelのCB情報を出力

meta.data <- read.csv(meta.data_input, header = TRUE)
CBsplit <- str_split(as.character(meta.data$X), pattern="_", simplify = TRUE)
meta.data$CB <- as.numeric(CBsplit[,3])


##Experimentに相当するデータの抽出
#CD8 (全細胞をCD8+ T cellとして解析)

meta_data_endo <- meta.data

name_CB_array <- c("endo")
CB_list <- list(meta_data_endo)

setwd( paste(".", "filtered2", sep = "/") )
files_TCR <- list.files(pattern="CB.TR")

##Cell Barcodeを用いたfiltering
for (file.name1 in files_TCR){
  for (i in 1:length(name_CB_array)){
    
    #解析するcell subsetの決定
    meta_data_CB <- CB_list[[i]]
    name_CB <- name_CB_array[i]
    count_CB <- nrow(meta_data_CB)
    
    #TCR fileの読み込み
    data <- read.csv(file.name1, header = TRUE)
    names_TCR <- str_split(file.name1, pattern="\\.")

    #CBによるfiltering
    CB_endo_vec <- as.vector(meta_data_CB$CB)
    endo = data[data$CB %in% CB_endo_vec,]
    
    #遺伝子発現量の付加
    CB_endo_vec_TCR <- as.vector(endo$CB)
    meta_data_CB_TCR <- meta_data_CB[meta_data_CB$CB %in% CB_endo_vec_TCR,]
    endo <- endo[order(endo$CB),]
    meta_data_CB_TCR <- meta_data_CB_TCR[order(meta_data_CB_TCR$CB),]
    
    endo_out <- cbind(endo, meta_data_CB_TCR)
    
    #書き出し
    output.name <- paste(name_CB, names_TCR[[1]][1], names_TCR[[1]][2],
                         "count_th", names_TCR[[1]][4], "freq_th", names_TCR[[1]][6], "csv", sep=".")
    write.csv(endo_out, output.name)
    
    #Paring効率 / miss paring効率の出力
    paired <- subset(endo, A_cdr3nt != "UD" & B_cdr3nt != "UD")
    Pmela <- subset(endo, A_v == "TRAV7N-5" & A_j == "TRAJ40" & A_cdr3aa == "CAVNTGNYKYVF")
    Pmelb <- subset(endo, B_v == "TRBV14" & B_j == "TRBJ1-6" & B_cdr3aa == "CASSFHRDYNSPLYF")
    Pmelab <- subset(endo, A_v == "TRAV7N-5" & A_j == "TRAJ40" & A_cdr3aa == "CAVNTGNYKYVF" &
                       B_v == "TRBV14" & B_j == "TRBJ1-6" & B_cdr3aa == "CASSFHRDYNSPLYF")
    
    num.total <- nrow(endo)
    num.paired <- nrow(paired)
#    num.Pmel <- nrow(Pmelab)
#    num.misspaired <- nrow(Pmela) + nrow(Pmelb) - nrow(Pmelab)
    prop.paired <- paste(round(100*num.paired/count_CB, digits = 1), "%", sep="" ) 

    tail <- ("exp count.CB count.total count.paired paired.ratio\n") 
    out <- c(names_TCR[[1]][1], count_CB, num.total, num.paired, prop.paired, tail)
    cat(out, file = "output.txt", append = TRUE)
  }
}

filtered_tmp <- read.table("output.txt") 
out_filtered <- c(filtered_tmp[, c(1,2,3,4,5)])
names(out_filtered) <- c("exp", "count.CB", "count.total", "count.paired", "paired.ratio")
out_filtered <- as.data.frame(out_filtered)
```


```{r output_table1, echo=FALSE, cache=FALSE,dpi=300, out.width = 1000, out.height = 500, fig.align='center'}
DT::datatable(out_filtered, filter="none", width='100%',
          caption = 'Table 1:Filtered output',
          class = 'cell-border stripe',
          extensions = 'Buttons',
          options = list(pageLength=nrow(out_filtered), 
                         searching=FALSE, 
                         paging = FALSE,
                         info = FALSE,
                         lengthChange=TRUE, 
                         autoWidth = TRUE, 
                         columnDefs = list(list(className = 'dt-center', targets = 0:5)),
                         scrollX=F,
                         dom = 'Blfrtip',
                         buttons = list(list(extend = 'csv', filename= 'filtered.output'))
                         ))%>%
   DT::formatStyle(columns = c(1:ncol(out_filtered)), fontSize = 8)
```




```{r assemble, echo=FALSE, cache=FALSE, message=FALSE}

setwd( paste("..", dir_var, "filtered2", sep = "/") )


data <- read.csv(paste("endo.TRA_CB.TRB_CB.count_th.",count_th_,".freq_th.",freq_th_,".csv",sep=""), header = TRUE)

data_paired <- subset(data, data$A_cdr3nt != "UD" & data$B_cdr3nt != "UD")

if(clonotype=="ABnt"){
  data_paired$A_strict <- paste(data_paired$A_cdr3nt, data_paired$A_v, data_paired$A_j, sep="_")
  data_paired$B_strict <- paste(data_paired$B_cdr3nt, data_paired$B_v, data_paired$B_j, sep="_")
  data_paired$clone.id <- paste(data_paired$A_strict, data_paired$B_strict, sep="_")
}

if(clonotype=="ABaa"){
  data_paired$A_strict <- paste(data_paired$A_cdr3aa, data_paired$A_v, data_paired$A_j, sep="_")
  data_paired$B_strict <- paste(data_paired$B_cdr3aa, data_paired$B_v, data_paired$B_j, sep="_")
  data_paired$clone.id <- paste(data_paired$A_strict, data_paired$B_strict, sep="_")
}

if(clonotype=="Bnt"){
#  data_paired$A_strict <- paste(data_paired$A_cdr3nt, data_paired$A_v, data_paired$A_j, sep="_")
  data_paired$B_strict <- paste(data_paired$B_cdr3nt, data_paired$B_v, data_paired$B_j, sep="_")
  data_paired$clone.id <- data_paired$B_strict
}

if(clonotype=="Baa"){
#  data_paired$A_strict <- paste(data_paired$A_cdr3nt, data_paired$A_v, data_paired$A_j, sep="_")
  data_paired$B_strict <- paste(data_paired$B_cdr3aa, data_paired$B_v, data_paired$B_j, sep="_")
  data_paired$clone.id <- data_paired$B_strict
}

index <- unique(data_paired$clone.id)
total <- nrow(data_paired)

combined_table <- data.frame()

for (i in 1:length(index)){
  clone <- index[i]
  sub_clone <- subset(data_paired, clone.id == clone)
  count <- nrow(sub_clone)
  freq <- count/total
  tcrb_sel <- sub_clone[1,7:11]
  combine <- c(count, freq, tcrb_sel, clone)
  combine <- as.data.frame(combine)
  names(combine) = c( "count", "freq", "cdr3nt", "cdr3aa", "v", "j", "d", "clone")
  combined_table <- rbind(combined_table, combine)
}

combined_table <- combined_table[order(combined_table$count, decreasing = TRUE),]
write.csv(data_paired, paste("CellTable.paired.TCR.",clonotype,".csv",sep=""), row.names = FALSE)
write.table(combined_table, paste("scTCR_table.",clonotype,".txt",sep=""), row.names = F, sep="\t", quote = F)

```

--------

# 代表的なT細胞クローンの性質

TCRaおよびTCRbのCDR3のヌクレオチド配列をもとに、T細胞クローンを同定し、各クローンの総細胞数 (cell_count) と頻度 (freq) を計算した。

続いて、SCT解析における細胞のクラスタリング結果を利用し、T細胞クローンごとにそれぞれのクラスターに属する細胞数を集計した (clusterX)。

加えて、代表的なクローンについては、クローンを構成するT細胞の遺伝子発現プロファイルを次元圧縮して示した ("Figure" 内の左図)。加えて、各クローンを構成するT細胞のうち、single cell RNAseqで同定されたクラスターに属する細胞の割合を円グラフで表示している ("Figure" 内の右図)。

```{r BD_SCT_scTCR, echo=FALSE, cache=FALSE, message=FALSE, warning=FALSE, fig.show='hide', results='hide'}
setwd( paste("..", dir_var, sep = "/") )

createEmptyDf = function( nrow, ncol, colnames = c() ){
  data.frame( matrix( vector(), nrow, ncol, dimnames = list( c(), colnames ) ) )
}
set.seed(seed = 42)

dir.name <- "scTCR_analysis"
dir.create(dir.name)
dir.figures <- paste("figures.", clonotype, sep="")
dir.create(paste0("../",dir.output,"/",dir.figures))

sample.name = paste(rda_input,".rda",sep="")

#元データの読み込み
load(sample.name)

#scTCR, クローン情報の読み込み
clone_id_table <- read.csv(paste("./filtered2/",paste("CellTable.paired.TCR.",clonotype,".csv",sep=""),sep=""), header = TRUE)

clone.id.list <- clone_id_table$`clone.id`
names(clone.id.list)=clone_id_table$X

#seu = UpdateSeuratObject(object = seu)
seu <- AddMetaData(object = seu, metadata = clone.id.list, col.name = "clone.id")

meta.data <- as.data.frame(seu@meta.data)

write.csv(meta.data, paste0("../",dir.output,"/","Others","/","meta.data.afterscTCR.",clonotype,".csv"), row.names = TRUE)


##############################################################################
#クローンレベルでの性質の集計

tmp = clone_id_table %>% group_by(`clone.id`) %>%
  dplyr::summarise(count = n()) %>%
  dplyr::arrange(desc(count))
tmp = as.data.frame(tmp)
tmp[,3]=tmp[,2]/sum(tmp[,2])
colnames(tmp)=c("ntSeq_TRA_TRB_freq", "CloneCount", "CloneFreq")
tmp <- tmp[order(tmp$ntSeq_TRA_TRB_freq),]

tmp = cbind(tmp, orig.ident=rep("CD8T", nrow(tmp)))
temp_count <- as.vector(tmp$CloneCount)
temp_freq <- as.vector(tmp$CloneFreq)

#TCRクローンごとに頻度を集計
tmp_out = table(seu@active.ident, seu@meta.data$clone.id)
tmp_out = t(tmp_out)
cluster_num<-ncol(tmp_out) #クラスター数
file.name=paste(dir.name, "clone_within_cluster.txt", sep='/')
write.table(tmp_out, file.name, row.names=T, col.names=T, sep="\t", quote=F)
tmp_out2 <- read.table(file.name, header = TRUE)
tmp_out2$names <- row.names(tmp_out2)
tmp_out2 <- tmp_out2[order(tmp_out2$names),]

tmp_out3 <- cbind(temp_count, temp_freq, tmp_out2)
write.table(tmp_out3, file.name, row.names=F, col.names=T, sep="\t", quote=F)
out_BD <- tmp_out3[order(tmp_out3$temp_freq, decreasing=T),]

out_BD_all <- out_BD

#rank <- rank(-out_BD$temp_freq)
rank <-  1:nrow(out_BD)

out_BD <- cbind(rank, out_BD)

out_BD$temp_freq <- paste(round(100*out_BD$temp_freq, digits = 2), "%", sep="" )

out_BD <- data.frame(out_BD)

out_BD$rank <- paste("Top",out_BD$rank, sep=" ")

cluster_names<-rep("", cluster_num)

for (i in 1:cluster_num) {
  cluster_names[i]<-paste("cluster",formatC(i-1,width=2,flag="0"))
}

names(out_BD) <- c("name","cell_count", "freq", cluster_names, "Sequence")

out_BD2 <- out_BD[1:10,]

out_BD_all_table <- out_BD

out_BD <- out_BD[out_BD$cell_count>10,]

file.name=paste0("../",dir.output,"/",rda_input, "scTCRmerged.", clonotype, ".rda")
save(seu, file=file.name)

write.table(as.matrix(GetAssayData(object = seu)), 
            paste0('../',dir.output,'/','scTCRmerged.',clonotype,'.txt'), 
            sep = ',', row.names = T, col.names = T, quote = F)

cluster_vec<-1:cluster_num
vec_tmp<- rep(1,cluster_num)


pie_charts <- as.data.frame(t(out_BD[cluster_vec + 3*rep(1,cluster_num)]))
names <- names(pie_charts)

pie_charts_all <- colSums(out_BD_all[cluster_vec + 2*rep(1,cluster_num)])

#namesに"/"が入っていたらエラーになるので変換
names_tmp<-gsub("/", "__", names)


data <- data.frame(
  group=cluster_names,
  value=as.vector(pie_charts_all)
)
data <- data %>% 
  arrange(desc(group)) %>%
  mutate(prop = value / sum(data$value) *100) %>%
  mutate(ypos = cumsum(prop)- 0.5*prop )

#active.identでdimplotするために新たにmBC2を作成
seu2 = AddMetaData(seu, metadata = as.character(as.numeric(seu@active.ident)-1), col.name = "active.ident")

##Dimplotに使用する色の抜き出し
p_color <- DimPlot(object = seu2, group.by = "active.ident", reduction = "tsne", pt.size = 1)
orig_ident_build = ggplot2::ggplot_build(p_color)
orig_ident_build = orig_ident_build$data[[1]]
orig_ident_build2 <- orig_ident_build
orig_ident_build =  orig_ident_build[order(orig_ident_build$group), ]
ident.cols <- unique(orig_ident_build$colour) # Get a vector of unique colors
names(ident.cols)<-c(0:(max(as.numeric(seu2@active.ident))-1))

##Dimplotに使用するx,y軸情報の抜き出し
xmin <- min(orig_ident_build$x)-2
xmax <- max(orig_ident_build$x)+2
ymin <- min(orig_ident_build$y)-2
ymax <- max(orig_ident_build$y)+2
x_label_min <- floor(xmin/10)*10
x_label_max <- ceiling(xmax/10)*10
y_label_min <- floor(ymin/10)*10
y_label_max <- ceiling(ymax/10)*10


p <-ggplot(data, aes(x="", y=prop, fill=group)) +
  geom_bar(stat="identity", width=1, color="white") +
  coord_polar("y", start=0, direction = -1) +
  theme_void() +
  geom_text(data = (. %>% filter(prop > 0)), aes(label = paste0(round(prop, digit=1), "%")), position = position_stack(vjust = 0.5), size=4)+
  scale_fill_manual(values=as.vector(ident.cols))
ggsave(file = paste0("../",dir.output,"/",dir.figures,"/","all_pie.png"), plot = p, dpi = 100, width = 6.4, height = 4.8)


p1 <- DimPlot(object = seu2, group.by = "active.ident", 
              label = FALSE,
              reduction = "tsne",
              cols = ident.cols,
              pt.size = 1) + NoLegend() +
    #軸の範囲が常に全体のプロットと一致するようにする
    scale_x_continuous(limits = c(xmin,xmax), breaks= seq(x_label_min,x_label_max,10)) +
    scale_y_continuous(limits = c(ymin,ymax), breaks= seq(y_label_min,y_label_max,10)) +
    theme(axis.title = element_blank(), 
          axis.text = element_blank())

file.name <- paste0("../",dir.output,"/",dir.figures,"/","all_pl.png")
save_plot(file = file.name, plot(p1), device="png", 
          units="in", dpi = 100, base_width = 4.8, base_height = 4.8, limitsize=FALSE)
fig1 <- image_read(file.name) 
fig2 <- image_read(paste0("../",dir.output,"/",dir.figures,"/","all_pie.png"))
img <- c(fig1,fig2)
img <- image_append(image_scale(img, "1000x480")) 
invisible(image_write(img, path = paste0("../",dir.output,"/",dir.figures,"/","all.png"), format = "png") )


tmp_raw <- as.vector(row.names(meta.data))

Pie_chart <- function(num){
  data <- data.frame(
    group=cluster_names,
    value=as.vector(pie_charts[,num])
  )
  data <- data %>% 
    arrange(desc(group)) %>%
    mutate(prop = value / sum(data$value) *100) %>%
    mutate(ypos = cumsum(prop)- 0.5*prop )

  p <-ggplot(data, aes(x="", y=prop, fill=group)) +
    geom_bar(stat="identity", width=1, color="white") +
    coord_polar("y", start=0, direction = -1) +
    theme_void() +
    geom_text(data = (. %>% dplyr::filter(prop > 0)), aes(label = paste0(round(prop, digit=1), "%")), position = position_stack(vjust = 0.5), size=4)+
    scale_fill_manual(values=as.vector(ident.cols))
  ggsave(file = paste0("../",dir.output,"/",dir.figures,"/",names_tmp[num],"_pie.png"), plot = p, dpi = 100, width = 6.4, height = 4.8)
  
##Scatter
  clone_name <- names[num]
  data_sub <- dplyr::filter(meta.data, clone.id == clone_name)
  clone.id <- as.vector(row.names(data_sub))
  tmp2 = tmp_raw %in% clone.id
  p1 <- DimPlot(object = seu2, group.by = "active.ident", 
                label = FALSE,
                cells =  row.names(meta.data)[tmp2],
                reduction = "tsne", 
                cols= ident.cols,
                pt.size = 2.5) + NoLegend() +
    #軸の範囲が常に全体のプロットと一致するようにする
    scale_x_continuous(limits = c(xmin,xmax), breaks= seq(x_label_min,x_label_max,10)) +
    scale_y_continuous(limits = c(ymin,ymax), breaks= seq(y_label_min,y_label_max,10)) +
    theme(axis.title = element_blank(), 
          axis.text = element_blank())
  
  #clne_nameから"/"を除く
  clone_name<-gsub("/", "__", clone_name)
  file.name <- paste0("../",dir.output,"/",dir.figures, "/" ,clone_name, "_pl.png")
  save_plot(file = file.name, plot(p1), device="png", 
            units="in", dpi = 100, base_width = 4.8, base_height = 4.8, limitsize=FALSE)
  fig1 <- image_read(file.name) 
  fig2 <- image_read(paste0("../",dir.output,"/",dir.figures, "/", clone_name,"_pie.png"))
  img <- c(fig1,fig2)
  img <- image_append(image_scale(img, "1000x480")) 
  invisible(image_write(img, path = paste0("../",dir.output,"/",dir.figures,"/",clone_name,".png"), format = "png"))
}

cl <- makeCluster(cores)
registerDoParallel(cl)
invisible(foreach(i = 1:ncol(pie_charts) , .packages=c("stringr", "ggplot2","dplyr","Seurat","cowplot","magick")) %dopar% {
  Pie_chart(i)
  })
stopCluster(cl)

system(paste("rm ", "../",dir.output,"/",dir.figures,"/*_pie.png", sep="") )
system(paste("rm ", "../",dir.output,"/",dir.figures,"/*_pl.png", sep=""))
system(paste("rm ", "./","Rplots.pdf", sep=""))

```


```{r, echo=FALSE, out.width="70%", fig.align="center", fig.cap="Scatter plot and Pie charts."}
knitr::include_graphics(paste("./",dir.figures,"/all.png", sep="" ))
```

全細胞でのScatter plotとPie chartsの.pngファイルは，"all.png" として，"/Output.files/`r paste("figures.", clonotype, sep="")`/"に保存した．

```{r output_table2, echo=FALSE, cache=FALSE,dpi=300, out.width = 1000, out.height = 500, fig.align='center'}

Figures <- rep('figure', nrow(out_BD_all_table))
for (i in 0:nrow(out_BD_all_table)) {
  if(i<=nrow(out_BD)){
    Figures[i] <- paste("<a href='./",dir.figures,"/", out_BD$Sequence[i], ".png'> Figure </a>", sep="")
  }else{
    Figures[i] <- paste("-")
  }
}

invisible(Figures < as.data.frame(Figures))

DT::datatable(cbind(Figures,out_BD_all_table), filter="none", width='100%',rownames = FALSE,
          caption = 'Table 2:BD output',
          escape=FALSE,
          class = 'cell-border stripe',
          extensions = c('Buttons','FixedColumns'),
          options = list(pageLength=nrow(out_BD_all_table), 
                         searching=FALSE, 
                         paging = FALSE,
                         info = FALSE,
                         lengthChange=TRUE, 
                         autoWidth = TRUE, 
                         scrollX = TRUE,
                         scrollY = "370px",
                         columnDefs = list(list(className = 'dt-center', targets = 0:12)),
                         dom = 'Blfrtip',
                         buttons = list(list(extend = 'csv', filename= 'clone_within_cluster'))
                         ))%>%
   DT::formatStyle(columns = c(1:ncol(out_BD_all_table)), fontSize = 8)
```



```{r output, echo=FALSE, cache=FALSE, message=FALSE, warning=FALSE}


dir.create("TRA")
dir.create("TRB")

#invisible(file.copy(paste0("../", dir_var, "/", "figures.ABnt"), paste0("./","Output.files"), recursive=TRUE))
#unlink(paste0("../", dir_var, "/", "figures.ABnt"), recursive = TRUE)


#lf_from <- paste("../", dir_var , paste("/meta.data.afterscTCR.",clonotype,".csv",sep=""), sep="")
#lf_to   <- "./Others/"
#invisible(file.copy(from=lf_from, to=lf_to))   #ファイルをコピー


lf_from <- list.files(paste("../", dir_var ,"/filtered2/",sep=""), full.names=T)
lf_to   <- "./Others/"
invisible(file.copy(from=lf_from, to=lf_to))   #ファイルをコピー

lf_from <- list.files(paste("../", dir_var ,"/filtered2/TRA/",sep=""), pattern="tiff", full.names=T)
lf_from <- append(lf_from,paste("../", dir_var ,"/filtered2/TRA/TRA_CB.txt",sep=""))
lf_to   <- "./TRA/"
invisible(file.copy(from=lf_from, to=lf_to))   #ファイルをコピー


lf_from <- list.files(paste("../", dir_var ,"/filtered2/TRB/",sep=""), pattern="tiff", full.names=T)
lf_from <- append(lf_from,paste("../", dir_var ,"/filtered2/TRB/TRB_CB.txt",sep=""))
lf_to   <- "./TRB/"
invisible(file.copy(from=lf_from, to=lf_to))   #ファイルをコピー

setwd("../")
#command <- sprintf(paste('tar -jcvf dLN.tar.bz2 dLN \\;', sep=""))
#system(command)
#unlink("dLN", recursive = TRUE)

```

各クローンごとのScatter plot, Pie chartsの.pngファイルは，"/Output.files/`r paste("figures.", clonotype, sep="")`."に保存した．

また，scTCRの結果を反映させたSeurat objectファイル (`r paste0(rda_input, "scTCRmerged.", clonotype, ".rda")`)は，"/Output.files/"に保存した．
